<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Peter Schmidtke">
<meta name="dcterms.date" content="2023-04-03">
<meta name="description" content="First post in a row on setting up a proper benchmark dataset for binding site comparison algorithm evaluation.">

<title>Spinning coral - Binding Site Comparison Benchmarks - I - Binding sites on the same protein</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-9LJZZ6RBGN"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-9LJZZ6RBGN', { 'anonymize_ip': true});
</script>
<script type="module" src="../../site_libs/quarto-ojs/quarto-ojs-runtime.js"></script>
<link href="../../site_libs/quarto-ojs/quarto-ojs.css" rel="stylesheet">


<link rel="stylesheet" href="../../styles.scss">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Spinning coral</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/pschmidtke" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://masto.ai/(pschmidtke?)" rel="" target=""><i class="bi bi-mastodon" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#summary" id="toc-summary" class="nav-link active" data-scroll-target="#summary">Summary</a></li>
  <li><a href="#the-target-binding-site" id="toc-the-target-binding-site" class="nav-link" data-scroll-target="#the-target-binding-site">The target binding site</a></li>
  <li><a href="#hsp90" id="toc-hsp90" class="nav-link" data-scroll-target="#hsp90">HSP90</a>
  <ul class="collapse">
  <li><a href="#why-hsp90-as-a-first-example" id="toc-why-hsp90-as-a-first-example" class="nav-link" data-scroll-target="#why-hsp90-as-a-first-example">Why HSP90 as a first example</a></li>
  <li><a href="#domain-architecture" id="toc-domain-architecture" class="nav-link" data-scroll-target="#domain-architecture">Domain architecture</a></li>
  <li><a href="#the-topology-of-the-atp-binding-site-n-ter" id="toc-the-topology-of-the-atp-binding-site-n-ter" class="nav-link" data-scroll-target="#the-topology-of-the-atp-binding-site-n-ter">The topology of the ATP binding site (N-ter)</a></li>
  </ul></li>
  <li><a href="#establishing-a-reference-set" id="toc-establishing-a-reference-set" class="nav-link" data-scroll-target="#establishing-a-reference-set">Establishing a reference set</a>
  <ul class="collapse">
  <li><a href="#same-sequence-conformation" id="toc-same-sequence-conformation" class="nav-link" data-scroll-target="#same-sequence-conformation">1. Same Sequence &amp; conformation</a>
  <ul class="collapse">
  <li><a href="#defining-the-binding-site" id="toc-defining-the-binding-site" class="nav-link" data-scroll-target="#defining-the-binding-site">Defining the binding site</a></li>
  <li><a href="#gathering-all-pdb-codes-containing-the-site" id="toc-gathering-all-pdb-codes-containing-the-site" class="nav-link" data-scroll-target="#gathering-all-pdb-codes-containing-the-site">Gathering all PDB codes containing the site</a></li>
  <li><a href="#cluster-by-conformation" id="toc-cluster-by-conformation" class="nav-link" data-scroll-target="#cluster-by-conformation">Cluster by conformation</a></li>
  <li><a href="#automating-things" id="toc-automating-things" class="nav-link" data-scroll-target="#automating-things">Automating things</a></li>
  <li><a href="#wrap-up" id="toc-wrap-up" class="nav-link" data-scroll-target="#wrap-up">Wrap-up</a></li>
  <li><a href="#comparison-to-existing-literature" id="toc-comparison-to-existing-literature" class="nav-link" data-scroll-target="#comparison-to-existing-literature">Comparison to existing literature</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Binding Site Comparison Benchmarks - I - Binding sites on the same protein</h1>
  <div class="quarto-categories">
    <div class="quarto-category">binding site</div>
    <div class="quarto-category">pocket</div>
    <div class="quarto-category">cavity</div>
    <div class="quarto-category">pocket comparison</div>
    <div class="quarto-category">structure-based drug design</div>
  </div>
  </div>

<div>
  <div class="description">
    First post in a row on setting up a proper benchmark dataset for binding site comparison algorithm evaluation.
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Peter Schmidtke </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 3, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="summary" class="level1 page-columns page-full">
<h1>Summary</h1>
<p>In this article Iâ€™ll go through the overall thought process of setting up a benchmark set. For the sake of simplicity Iâ€™ll do it for a single target to evaluate the performance of a binding site comparison algorithm. Iâ€™ll try to provide code whenever possible against public resources or snippets if a bit of scripting is needed. I am describing my thought process here. I will for sure run into dead ends (I wonâ€™t describe all of these <span class="emoji" data-emoji="grimacing">ðŸ˜¬</span>) and as a result Iâ€™ll have to make some decisions &amp; put other things aside. In order to keep track of things I put aside Iâ€™ll use a side-note throughout the post to track all of these &amp; make a summary of how to address or amend them later.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Things to keep in mind</p>
<ul>
<li>listing goes here</li>
</ul>
</div></div></section>
<section id="the-target-binding-site" class="level1">
<h1>The target binding site</h1>
<p>In order to show the overall process, letâ€™s start with a well studied target &amp; relatively easy target, HSP90. A usual guinea pig since my PhD in Xavier Barrilâ€™s lab, but itâ€™s a good example to show without exploring too much in super large protein families in the beginning. I aim to apply the process outlined here later to a thrombin, a trypsin like serine protease, which will likely involve some adjustments.</p>
</section>
<section id="hsp90" class="level1">
<h1>HSP90</h1>
<p>The heat shock protein 90 is a rather abundant protein in the cell and helps during protein folding of not yet formed proteins, or protects already folded proteins from external stress (thermal stress for instance) - thus HSP = Heat Shock Protein. Letâ€™s start with <a href="https://www.uniprot.org/uniprotkb/P07900/entry">HSP90 alpha from the homo sapiens</a>, even though they are common among eucaryotes. Several kinases are dependent (activated) on HSP90, especially those acting as hub (<span class="citation" data-cites="Citri2006">Citri et al. (<a href="#ref-Citri2006" role="doc-biblioref">2006</a>)</span>). This is one of the reasons why HSP90 has been investigated as potential drug target for treatment of several forms of cancer. The action of HSP is dependent on ATP and the dephosphorylation of this molecule &amp; the ATP binding site is located on a particular are in the N-terminal part of HSP90. This will be the binding site to focus on here.</p>
<section id="why-hsp90-as-a-first-example" class="level2">
<h2 class="anchored" data-anchor-id="why-hsp90-as-a-first-example">Why HSP90 as a first example</h2>
<ul>
<li>HSP90 has a lot of structures available in the public domain.</li>
<li>Itâ€™s not part of any gigantic protein family (kinases, GPCRâ€™s etc), which keeps the initial comparison space to cover a bit smaller.</li>
<li>The fold of the protein is still conserved among several other proteins, so thereâ€™s matter for detecting expected similarities</li>
<li>It binds ATP, such as a lot of proteins in nature - which is interesting for the if I bind the same molecule I must be similar conundrum.</li>
<li>The binding site can undergo important conformational changes, which is good to evaluate sensity on conformation</li>
<li>Water plays a very important role upon binding of small molecules into the ATP binding site</li>
</ul>
</section>
<section id="domain-architecture" class="level2">
<h2 class="anchored" data-anchor-id="domain-architecture">Domain architecture</h2>
<p>HSP90 alpha human is composed of two domains: - the N-terminal Histidine kinase, DNA gyrase B and HSP90-like ATPase domain (ranging from amino acid 40 to 193) - the C-terminal HSP90 protein domain (196-714)</p>
<p>The ATP binding site of interest is on the N-terminal part and this is the part that you have the most crystal strucutres for in the RCSB today. A full length Alphafold model available in the public domain. NB: there appears to be another ATP binding site on the C-terminal part, that is only accessible when activated - so interesting to maybe look out for that one as well (<span class="citation" data-cites="Soti2003">Soti et al. (<a href="#ref-Soti2003" role="doc-biblioref">2003</a>)</span>).</p>
</section>
<section id="the-topology-of-the-atp-binding-site-n-ter" class="level2">
<h2 class="anchored" data-anchor-id="the-topology-of-the-atp-binding-site-n-ter">The topology of the ATP binding site (N-ter)</h2>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb1" data-startfrom="66" data-source-offset="0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 65;"><span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a><span class="co">// Create drawing area</span></span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a>divNGL <span class="op">=</span> <span class="fu">html</span><span class="vs">`&lt;div style="width:800px;height:600px;position:relative"&gt;&lt;/div&gt;`</span><span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-hsp-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div id="ojs-cell-1" data-nodetype="declaration">

</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;1: Overview of N-terminal domain of human HSP90, canonical isoform</figcaption><p></p>
</figure>
</div>
</div>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb2" data-startfrom="75" data-source-offset="-0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 74;"><span id="cb2-75"><a href="#cb2-75" aria-hidden="true" tabindex="-1"></a>NGL <span class="op">=</span> <span class="pp">require</span>(<span class="st">"ngl@next"</span>)<span class="op">;</span></span>
<span id="cb2-76"><a href="#cb2-76" aria-hidden="true" tabindex="-1"></a><span class="co">// Create trajectory object</span></span>
<span id="cb2-77"><a href="#cb2-77" aria-hidden="true" tabindex="-1"></a>stage <span class="op">=</span> <span class="kw">new</span> NGL<span class="op">.</span><span class="fu">Stage</span>(divNGL<span class="op">,</span> { <span class="dt">backgroundColor</span><span class="op">:</span> <span class="st">"black"</span><span class="op">,</span><span class="dt">cameraType</span><span class="op">:</span> <span class="st">"orthographic"</span>})<span class="op">;</span></span>
<span id="cb2-78"><a href="#cb2-78" aria-hidden="true" tabindex="-1"></a>pdbString <span class="op">=</span> <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"3t0z.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">;</span></span>
<span id="cb2-79"><a href="#cb2-79" aria-hidden="true" tabindex="-1"></a>structure <span class="op">=</span> <span class="cf">await</span> stage<span class="op">.</span><span class="fu">loadFile</span>(pdbString<span class="op">,</span> {<span class="dt">ext</span><span class="op">:</span> <span class="st">"pdb"</span><span class="op">,</span> <span class="dt">asTrajectory</span><span class="op">:</span> <span class="kw">false</span>})</span>
<span id="cb2-80"><a href="#cb2-80" aria-hidden="true" tabindex="-1"></a>structure<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"cartoon"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"protein"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"residueindex"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-2" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-3" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-4" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-5" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb3" data-startfrom="81" data-source-offset="-360"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 80;"><span id="cb3-81"><a href="#cb3-81" aria-hidden="true" tabindex="-1"></a>structure<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"ball+stick"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"hetero and _C"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"yellow"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-6" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb4" data-startfrom="82" data-source-offset="-449"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 81;"><span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a>structure<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"ball+stick"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"hetero"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"element"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-7" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb5" data-startfrom="83" data-source-offset="-532"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 82;"><span id="cb5-83"><a href="#cb5-83" aria-hidden="true" tabindex="-1"></a>structure<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"licorice"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"47-55 or 91 or 93 or 96-97 or 98 or 102 or 106 or 107 or 112 or 132-139 or 150 or 162 or or 186"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"element"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-8" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb6" data-startfrom="84" data-source-offset="-702"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 83;"><span id="cb6-84"><a href="#cb6-84" aria-hidden="true" tabindex="-1"></a>structure<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"contact"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"47-55 or 91 or 93 or 96-97 or 98 or 102 or 106 or 107 or 112 or 132-139 or 150 or 162 or or 186 or ligand or water"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"element"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-9" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb7" data-startfrom="85" data-source-offset="-890"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 84;"><span id="cb7-85"><a href="#cb7-85" aria-hidden="true" tabindex="-1"></a>structure<span class="op">.</span><span class="fu">autoView</span>()<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-2-10" data-nodetype="expression">

</div>
</div>
</div>
</div>
<p>The binding site is composed of a section containing the adenin moety which is characterized by a beta sheet at the bottom of the site, and two helices lining the site. The sidechains exposed to the binding site lumen are globally hydrophobic, a part from the very important aspartate 93, which is interacting directly with the adenine moeity. The adenin moeity is sourrounded by water molecules and several of these watersare important hallmarks of several HSP90 binders. The ribose moeity is not forming any H-bonds with the protein itself but has hydroxyls oriented towards the solvent. The ether of the ribose is orented towards the valine 107, adjacent to a rather hydrophobic part of the pocket coated by Y139, F138 and W162. The triphospate is solvent exposed and interacting with a small helix-loop-helix motive which, as we will probably see a bit later part of the more mobile regions of the binding site.</p>
</section>
</section>
<section id="establishing-a-reference-set" class="level1 page-columns page-full">
<h1>Establishing a reference set</h1>
<p>Now we have defined a binding site of interest, letâ€™s try to establish first a few obvious scenarios one might want to cover with a pocket comparison method. In my previous post I stated that the principle use case for binding site comparison methods that Iâ€™m focusing on is large scale comparison or screening. The underlying use cases that Iâ€™m mainly interested in are NOT protein function prediction, but rather the prediction of potential counter targets, or extracting bound ligands from related binding sites to inform structurally during my compound design cycle.</p>
<p>Letâ€™s consider that we are working on a drug design project on the HSP90 N-terminal ATP binding site and we have our favourite structure of HSP90 as starting point, <a href="https://www.rcsb.org/structure/4CWR">i.e.&nbsp;4cwr</a>.</p>
<p>If I want to find similar binding sites to my query binding site vs all known/putative binding sites, what hits Iâ€™d expect to get first in the hitlist ?:</p>
<ol type="1">
<li>the ATP binding site of other HSP90 alpha human structures with the same overall conformation (more or less) - same sequence &amp; conformation</li>
<li>the ATP binding site of other HSP90 human isoforms with a similar conformation - likely locally identical sequences &amp; conformation</li>
<li>the ATP binding site of HSP90 sequences from other species with a similar conformation - locally very similar sequences</li>
<li>all of the above but with a bit different conformations - conformations</li>
<li>the ATP binding site of close homologs (sequence - families etc) to the query structure - similar sequences</li>
<li>the ATP binding sites of proteins sharing the same fold as HSP90 - same fold</li>
<li>nucleotide binding sites with similar interaction patterns but dissimar fold - same interactions</li>
<li>all binding sites binding ATP must be similar (provocative on purpose â€¦) - same ligand</li>
<li>unexpected &amp; unrelated / unknown similarities - nightmare</li>
</ol>
<p>These first five give a graduation up to which level another ATP binding site could be potentially close to the HSP90 binding site. These are the obvious clusters of sequences, structures &amp; conformations one would expect to find. As a result, one can use this type of graduation also for validating binding site comparison methods. One major difference with the setting Iâ€™m laying out here, is that the background data encompasses the full RCSB PDB structures containing all ligand binding sites + putative binding sites (empty clefts). This sets the approach Iâ€™m suggesting into stark contrast with previous benchmark sets. Such previous sets were classically composed of a list of expected matches of binding site pairs and decoys (expected mismatches). As Vincent Le Guilloux states it fairly regularly, if you can avoid a threshold effect, avoid it! This is exactly such a case where a discrete split between a match &amp; a mismatch is typically introduced. As a result my background data (what one usually calls a decoy) is the full pocketome and Iâ€™ll try to use metrics of success that measure, how many of the potentially expected hits are found before a bulk of less expected hits &amp; why.</p>
<p>In the subsequent sections Iâ€™ll go through all the painful steps to create the dataset corresponding the very first &amp; the easiest steps. Iâ€™ll cover 2 to 5 in different posts, as already this one here is getting fairly long &amp; complex. The situation 7 &amp; 8 are a bit more tricky to set up. As for 9 - thatâ€™s the big issue with binding site comparison benchmarking - you donâ€™t know until you know, but Iâ€™ll try to do a bit of my homework on that as well!</p>
<section id="same-sequence-conformation" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="same-sequence-conformation">1. Same Sequence &amp; conformation</h2>
<p>Alright, here we go â€¦ the same sequence is the easiest case of all of them &amp; several previous studies included a selection of structures, but as youâ€™ll see even here to do things properly it gets quickly tricky. The following script will cover the required steps: - define the sets of residues of our reference binding site - gather all structures (PDB codes) containing a resolved HSP90 alpha human N terminal domain - filter out structures with mutations on binding site residues compared to the wild type - get an all by all comparison of the binding sites (structurally speaking) which would allow for some rough clustering of conformations</p>
<section id="defining-the-binding-site" class="level3">
<h3 class="anchored" data-anchor-id="defining-the-binding-site">Defining the binding site</h3>
<p>On <a href="#fig-hsp-1">Figure&nbsp;1</a> we already displayed a set of amino-acids on the HSP90 structure sourrounding the binding site. Letâ€™s get a list defined by hand getting all residues:</p>
<div id="tbl-atp-binding-site-residues" class="anchored">
<table class="table">
<caption>Table&nbsp;1: ATP binding site residue selection</caption>
<thead>
<tr class="header">
<th>WT Sequence Position</th>
<th>Chain</th>
<th>Residue</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>E47</td>
<td>A</td>
<td>E47</td>
</tr>
<tr class="even">
<td>L48</td>
<td>A</td>
<td>L48</td>
</tr>
<tr class="odd">
<td>N51</td>
<td>A</td>
<td>N51</td>
</tr>
<tr class="even">
<td>S52</td>
<td>A</td>
<td>S52</td>
</tr>
<tr class="odd">
<td>D54</td>
<td>A</td>
<td>D54</td>
</tr>
<tr class="even">
<td>A55</td>
<td>A</td>
<td>A55</td>
</tr>
<tr class="odd">
<td>K58</td>
<td>A</td>
<td>K58</td>
</tr>
<tr class="even">
<td>D93</td>
<td>A</td>
<td>D93</td>
</tr>
<tr class="odd">
<td>G95</td>
<td>A</td>
<td>G95</td>
</tr>
<tr class="even">
<td>I96</td>
<td>A</td>
<td>I96</td>
</tr>
<tr class="odd">
<td>G97</td>
<td>A</td>
<td>G97</td>
</tr>
<tr class="even">
<td>M98</td>
<td>A</td>
<td>M98</td>
</tr>
<tr class="odd">
<td>N106</td>
<td>A</td>
<td>N106</td>
</tr>
<tr class="even">
<td>L107</td>
<td>A</td>
<td>L107</td>
</tr>
<tr class="odd">
<td>K112</td>
<td>A</td>
<td>K112</td>
</tr>
<tr class="even">
<td>I131</td>
<td>A</td>
<td>I131</td>
</tr>
<tr class="odd">
<td>G132</td>
<td>A</td>
<td>G132</td>
</tr>
<tr class="even">
<td>Q133</td>
<td>A</td>
<td>Q133</td>
</tr>
<tr class="odd">
<td>F134</td>
<td>A</td>
<td>F134</td>
</tr>
<tr class="even">
<td>G135</td>
<td>A</td>
<td>G135</td>
</tr>
<tr class="odd">
<td>V136</td>
<td>A</td>
<td>V136</td>
</tr>
<tr class="even">
<td>G137</td>
<td>A</td>
<td>G137</td>
</tr>
<tr class="odd">
<td>F138</td>
<td>A</td>
<td>F138</td>
</tr>
<tr class="even">
<td>Y139</td>
<td>A</td>
<td>Y139</td>
</tr>
<tr class="odd">
<td>T152</td>
<td>A</td>
<td>Y152</td>
</tr>
<tr class="even">
<td>G183</td>
<td>A</td>
<td>G183</td>
</tr>
<tr class="odd">
<td>T184</td>
<td>A</td>
<td>T184</td>
</tr>
<tr class="even">
<td>V186</td>
<td>A</td>
<td>V186</td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="gathering-all-pdb-codes-containing-the-site" class="level3 page-columns page-full">
<h3 class="anchored" data-anchor-id="gathering-all-pdb-codes-containing-the-site">Gathering all PDB codes containing the site</h3>
<p>Now we have the protein &amp; binding site of interest defined, we can gather all RCSB PDB structures containing that portion without mutations compared to the WT sequence. The uniprot accession for the human HSP90 is P07900 (HS90A_HUMAN) &amp; we will use the WT HSP90 alpha here as reference. NB: there is a known splice variant adding another 123 amino acids in front of ther canonical N-terminal part of the protein.</p>
<p>In order to retrieve all PDB structures of a particular protein sequences one can use various resources. The obvious one to start with would be uniprot, ncbi or the RCSB itself.</p>
<p>From uniprot we can use something the ID mapping service for instance (I go for python directly as the API runs an asynchroneous job):</p>
<div class="cell" data-execution_count="1">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Code adapted from here: https://www.uniprot.org/help/id_mapping</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> re</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> time</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> json</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> zlib</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> urllib.parse <span class="im">import</span> urlparse, parse_qs, urlencode</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> requests</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> requests.adapters <span class="im">import</span> HTTPAdapter, Retry</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>POLLING_INTERVAL <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>API_URL <span class="op">=</span> <span class="st">"https://rest.uniprot.org"</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>retries <span class="op">=</span> Retry(total<span class="op">=</span><span class="dv">5</span>, backoff_factor<span class="op">=</span><span class="fl">0.25</span>, status_forcelist<span class="op">=</span>[<span class="dv">500</span>, <span class="dv">502</span>, <span class="dv">503</span>, <span class="dv">504</span>])</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>session <span class="op">=</span> requests.Session()</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>session.mount(<span class="st">"https://"</span>, HTTPAdapter(max_retries<span class="op">=</span>retries))</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> check_response(response):</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>        response.raise_for_status()</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span> requests.HTTPError:</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(response.json())</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>        <span class="cf">raise</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> submit_id_mapping(from_db, to_db, ids):</span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>    request <span class="op">=</span> requests.post(</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>        <span class="ss">f"</span><span class="sc">{</span>API_URL<span class="sc">}</span><span class="ss">/idmapping/run"</span>,</span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>        data<span class="op">=</span>{<span class="st">"from"</span>: from_db, <span class="st">"to"</span>: to_db, <span class="st">"ids"</span>: <span class="st">","</span>.join(ids)},</span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>    check_response(request)</span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> request.json()[<span class="st">"jobId"</span>]</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_next_link(headers):</span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>    re_next_link <span class="op">=</span> re.<span class="bu">compile</span>(<span class="vs">r'&lt;(.+)&gt;; rel="next"'</span>)</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">"Link"</span> <span class="kw">in</span> headers:</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>        match <span class="op">=</span> re_next_link.match(headers[<span class="st">"Link"</span>])</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> match:</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> match.group(<span class="dv">1</span>)</span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> check_id_mapping_results_ready(job_id):</span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> <span class="va">True</span>:</span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a>        request <span class="op">=</span> session.get(<span class="ss">f"</span><span class="sc">{</span>API_URL<span class="sc">}</span><span class="ss">/idmapping/status/</span><span class="sc">{</span>job_id<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a>        check_response(request)</span>
<span id="cb8-49"><a href="#cb8-49" aria-hidden="true" tabindex="-1"></a>        j <span class="op">=</span> request.json()</span>
<span id="cb8-50"><a href="#cb8-50" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="st">"jobStatus"</span> <span class="kw">in</span> j:</span>
<span id="cb8-51"><a href="#cb8-51" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> j[<span class="st">"jobStatus"</span>] <span class="op">==</span> <span class="st">"RUNNING"</span>:</span>
<span id="cb8-52"><a href="#cb8-52" aria-hidden="true" tabindex="-1"></a>                <span class="bu">print</span>(<span class="ss">f"Retrying in </span><span class="sc">{</span>POLLING_INTERVAL<span class="sc">}</span><span class="ss">s"</span>)</span>
<span id="cb8-53"><a href="#cb8-53" aria-hidden="true" tabindex="-1"></a>                time.sleep(POLLING_INTERVAL)</span>
<span id="cb8-54"><a href="#cb8-54" aria-hidden="true" tabindex="-1"></a>            <span class="cf">else</span>:</span>
<span id="cb8-55"><a href="#cb8-55" aria-hidden="true" tabindex="-1"></a>                <span class="cf">raise</span> <span class="pp">Exception</span>(j[<span class="st">"jobStatus"</span>])</span>
<span id="cb8-56"><a href="#cb8-56" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb8-57"><a href="#cb8-57" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> <span class="bu">bool</span>(j[<span class="st">"results"</span>] <span class="kw">or</span> j[<span class="st">"failedIds"</span>])</span>
<span id="cb8-58"><a href="#cb8-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-59"><a href="#cb8-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-60"><a href="#cb8-60" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_batch(batch_response, file_format, compressed):</span>
<span id="cb8-61"><a href="#cb8-61" aria-hidden="true" tabindex="-1"></a>    batch_url <span class="op">=</span> get_next_link(batch_response.headers)</span>
<span id="cb8-62"><a href="#cb8-62" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> batch_url:</span>
<span id="cb8-63"><a href="#cb8-63" aria-hidden="true" tabindex="-1"></a>        batch_response <span class="op">=</span> session.get(batch_url)</span>
<span id="cb8-64"><a href="#cb8-64" aria-hidden="true" tabindex="-1"></a>        batch_response.raise_for_status()</span>
<span id="cb8-65"><a href="#cb8-65" aria-hidden="true" tabindex="-1"></a>        <span class="cf">yield</span> decode_results(batch_response, file_format, compressed)</span>
<span id="cb8-66"><a href="#cb8-66" aria-hidden="true" tabindex="-1"></a>        batch_url <span class="op">=</span> get_next_link(batch_response.headers)</span>
<span id="cb8-67"><a href="#cb8-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-68"><a href="#cb8-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-69"><a href="#cb8-69" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> combine_batches(all_results, batch_results, file_format):</span>
<span id="cb8-70"><a href="#cb8-70" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> file_format <span class="op">==</span> <span class="st">"json"</span>:</span>
<span id="cb8-71"><a href="#cb8-71" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> key <span class="kw">in</span> (<span class="st">"results"</span>, <span class="st">"failedIds"</span>):</span>
<span id="cb8-72"><a href="#cb8-72" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> key <span class="kw">in</span> batch_results <span class="kw">and</span> batch_results[key]:</span>
<span id="cb8-73"><a href="#cb8-73" aria-hidden="true" tabindex="-1"></a>                all_results[key] <span class="op">+=</span> batch_results[key]</span>
<span id="cb8-74"><a href="#cb8-74" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> file_format <span class="op">==</span> <span class="st">"tsv"</span>:</span>
<span id="cb8-75"><a href="#cb8-75" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> all_results <span class="op">+</span> batch_results[<span class="dv">1</span>:]</span>
<span id="cb8-76"><a href="#cb8-76" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb8-77"><a href="#cb8-77" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> all_results <span class="op">+</span> batch_results</span>
<span id="cb8-78"><a href="#cb8-78" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> all_results</span>
<span id="cb8-79"><a href="#cb8-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-80"><a href="#cb8-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-81"><a href="#cb8-81" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_id_mapping_results_link(job_id):</span>
<span id="cb8-82"><a href="#cb8-82" aria-hidden="true" tabindex="-1"></a>    url <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>API_URL<span class="sc">}</span><span class="ss">/idmapping/details/</span><span class="sc">{</span>job_id<span class="sc">}</span><span class="ss">"</span></span>
<span id="cb8-83"><a href="#cb8-83" aria-hidden="true" tabindex="-1"></a>    request <span class="op">=</span> session.get(url)</span>
<span id="cb8-84"><a href="#cb8-84" aria-hidden="true" tabindex="-1"></a>    check_response(request)</span>
<span id="cb8-85"><a href="#cb8-85" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> request.json()[<span class="st">"redirectURL"</span>]</span>
<span id="cb8-86"><a href="#cb8-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-87"><a href="#cb8-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-88"><a href="#cb8-88" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> decode_results(response, file_format, compressed):</span>
<span id="cb8-89"><a href="#cb8-89" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> compressed:</span>
<span id="cb8-90"><a href="#cb8-90" aria-hidden="true" tabindex="-1"></a>        decompressed <span class="op">=</span> zlib.decompress(response.content, <span class="dv">16</span> <span class="op">+</span> zlib.MAX_WBITS)</span>
<span id="cb8-91"><a href="#cb8-91" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> file_format <span class="op">==</span> <span class="st">"json"</span>:</span>
<span id="cb8-92"><a href="#cb8-92" aria-hidden="true" tabindex="-1"></a>            j <span class="op">=</span> json.loads(decompressed.decode(<span class="st">"utf-8"</span>))</span>
<span id="cb8-93"><a href="#cb8-93" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> j</span>
<span id="cb8-94"><a href="#cb8-94" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> file_format <span class="op">==</span> <span class="st">"tsv"</span>:</span>
<span id="cb8-95"><a href="#cb8-95" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> [line <span class="cf">for</span> line <span class="kw">in</span> decompressed.decode(<span class="st">"utf-8"</span>).split(<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>) <span class="cf">if</span> line]</span>
<span id="cb8-96"><a href="#cb8-96" aria-hidden="true" tabindex="-1"></a>        <span class="cf">elif</span> file_format <span class="op">==</span> <span class="st">"xlsx"</span>:</span>
<span id="cb8-97"><a href="#cb8-97" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> [decompressed]</span>
<span id="cb8-98"><a href="#cb8-98" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb8-99"><a href="#cb8-99" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> decompressed.decode(<span class="st">"utf-8"</span>)</span>
<span id="cb8-100"><a href="#cb8-100" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> file_format <span class="op">==</span> <span class="st">"json"</span>:</span>
<span id="cb8-101"><a href="#cb8-101" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> response.json()</span>
<span id="cb8-102"><a href="#cb8-102" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> file_format <span class="op">==</span> <span class="st">"tsv"</span>:</span>
<span id="cb8-103"><a href="#cb8-103" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> [line <span class="cf">for</span> line <span class="kw">in</span> response.text.split(<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>) <span class="cf">if</span> line]</span>
<span id="cb8-104"><a href="#cb8-104" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> file_format <span class="op">==</span> <span class="st">"xlsx"</span>:</span>
<span id="cb8-105"><a href="#cb8-105" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> [response.content]</span>
<span id="cb8-106"><a href="#cb8-106" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> response.text</span>
<span id="cb8-107"><a href="#cb8-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-108"><a href="#cb8-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-109"><a href="#cb8-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-110"><a href="#cb8-110" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> print_progress_batches(batch_index, size, total):</span>
<span id="cb8-111"><a href="#cb8-111" aria-hidden="true" tabindex="-1"></a>    n_fetched <span class="op">=</span> <span class="bu">min</span>((batch_index <span class="op">+</span> <span class="dv">1</span>) <span class="op">*</span> size, total)</span>
<span id="cb8-112"><a href="#cb8-112" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Fetched: </span><span class="sc">{</span>n_fetched<span class="sc">}</span><span class="ss"> / </span><span class="sc">{</span>total<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb8-113"><a href="#cb8-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-114"><a href="#cb8-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-115"><a href="#cb8-115" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_id_mapping_results_search(url):</span>
<span id="cb8-116"><a href="#cb8-116" aria-hidden="true" tabindex="-1"></a>    parsed <span class="op">=</span> urlparse(url)</span>
<span id="cb8-117"><a href="#cb8-117" aria-hidden="true" tabindex="-1"></a>    query <span class="op">=</span> parse_qs(parsed.query)</span>
<span id="cb8-118"><a href="#cb8-118" aria-hidden="true" tabindex="-1"></a>    file_format <span class="op">=</span> query[<span class="st">"format"</span>][<span class="dv">0</span>] <span class="cf">if</span> <span class="st">"format"</span> <span class="kw">in</span> query <span class="cf">else</span> <span class="st">"tsv"</span></span>
<span id="cb8-119"><a href="#cb8-119" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="st">"size"</span> <span class="kw">in</span> query:</span>
<span id="cb8-120"><a href="#cb8-120" aria-hidden="true" tabindex="-1"></a>        size <span class="op">=</span> <span class="bu">int</span>(query[<span class="st">"size"</span>][<span class="dv">0</span>])</span>
<span id="cb8-121"><a href="#cb8-121" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb8-122"><a href="#cb8-122" aria-hidden="true" tabindex="-1"></a>        size <span class="op">=</span> <span class="dv">500</span></span>
<span id="cb8-123"><a href="#cb8-123" aria-hidden="true" tabindex="-1"></a>        query[<span class="st">"size"</span>] <span class="op">=</span> size</span>
<span id="cb8-124"><a href="#cb8-124" aria-hidden="true" tabindex="-1"></a>    compressed <span class="op">=</span> (</span>
<span id="cb8-125"><a href="#cb8-125" aria-hidden="true" tabindex="-1"></a>        query[<span class="st">"compressed"</span>][<span class="dv">0</span>].lower() <span class="op">==</span> <span class="st">"true"</span> <span class="cf">if</span> <span class="st">"compressed"</span> <span class="kw">in</span> query <span class="cf">else</span> <span class="va">False</span></span>
<span id="cb8-126"><a href="#cb8-126" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb8-127"><a href="#cb8-127" aria-hidden="true" tabindex="-1"></a>    parsed <span class="op">=</span> parsed._replace(query<span class="op">=</span>urlencode(query, doseq<span class="op">=</span><span class="va">True</span>))</span>
<span id="cb8-128"><a href="#cb8-128" aria-hidden="true" tabindex="-1"></a>    url <span class="op">=</span> parsed.geturl()</span>
<span id="cb8-129"><a href="#cb8-129" aria-hidden="true" tabindex="-1"></a>    request <span class="op">=</span> session.get(url)</span>
<span id="cb8-130"><a href="#cb8-130" aria-hidden="true" tabindex="-1"></a>    check_response(request)</span>
<span id="cb8-131"><a href="#cb8-131" aria-hidden="true" tabindex="-1"></a>    results <span class="op">=</span> decode_results(request, file_format, compressed)</span>
<span id="cb8-132"><a href="#cb8-132" aria-hidden="true" tabindex="-1"></a>    total <span class="op">=</span> <span class="bu">int</span>(request.headers[<span class="st">"x-total-results"</span>])</span>
<span id="cb8-133"><a href="#cb8-133" aria-hidden="true" tabindex="-1"></a>    print_progress_batches(<span class="dv">0</span>, size, total)</span>
<span id="cb8-134"><a href="#cb8-134" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, batch <span class="kw">in</span> <span class="bu">enumerate</span>(get_batch(request, file_format, compressed), <span class="dv">1</span>):</span>
<span id="cb8-135"><a href="#cb8-135" aria-hidden="true" tabindex="-1"></a>        results <span class="op">=</span> combine_batches(results, batch, file_format)</span>
<span id="cb8-136"><a href="#cb8-136" aria-hidden="true" tabindex="-1"></a>        print_progress_batches(i, size, total)</span>
<span id="cb8-137"><a href="#cb8-137" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> results</span>
<span id="cb8-138"><a href="#cb8-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-139"><a href="#cb8-139" aria-hidden="true" tabindex="-1"></a>job_id <span class="op">=</span> submit_id_mapping(</span>
<span id="cb8-140"><a href="#cb8-140" aria-hidden="true" tabindex="-1"></a>    from_db<span class="op">=</span><span class="st">"UniProtKB_AC-ID"</span>, to_db<span class="op">=</span><span class="st">"PDB"</span>, ids<span class="op">=</span>[<span class="st">"P07900"</span>]</span>
<span id="cb8-141"><a href="#cb8-141" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-142"><a href="#cb8-142" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> check_id_mapping_results_ready(job_id):</span>
<span id="cb8-143"><a href="#cb8-143" aria-hidden="true" tabindex="-1"></a>    link <span class="op">=</span> get_id_mapping_results_link(job_id)</span>
<span id="cb8-144"><a href="#cb8-144" aria-hidden="true" tabindex="-1"></a>    results <span class="op">=</span> get_id_mapping_results_search(link)</span>
<span id="cb8-145"><a href="#cb8-145" aria-hidden="true" tabindex="-1"></a>    results_python <span class="op">=</span> json.loads(results[<span class="dv">0</span>])[<span class="st">"results"</span>]</span>
<span id="cb8-146"><a href="#cb8-146" aria-hidden="true" tabindex="-1"></a>    <span class="co"># print(results_python)</span></span>
<span id="cb8-147"><a href="#cb8-147" aria-hidden="true" tabindex="-1"></a>    up_pdbcodes<span class="op">=</span>[element[<span class="st">"to"</span>].lower() <span class="cf">for</span> element <span class="kw">in</span> results_python]</span>
<span id="cb8-148"><a href="#cb8-148" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(up_pdbcodes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Fetched: 328 / 328
['1byq', '1osf', '1uy6', '1uy7', '1uy8', '1uy9', '1uyc', '1uyd', '1uye', '1uyf', '1uyg', '1uyh', '1uyi', '1uyk', '1uyl', '1yc1', '1yc3', '1yc4', '1yer', '1yes', '1yet', '2bsm', '2bt0', '2bug', '2byh', '2byi', '2bz5', '2c2l', '2ccs', '2cct', '2ccu', '2fwy', '2fwz', '2h55', '2jjc', '2k5b', '2qf6', '2qfo', '2qg0', '2qg2', '2uwd', '2vci', '2vcj', '2wi1', '2wi2', '2wi3', '2wi4', '2wi5', '2wi6', '2wi7', '2xab', '2xdk', '2xdl', '2xds', '2xdu', '2xdx', '2xhr', '2xht', '2xhx', '2xjg', '2xjj', '2xjx', '2xk2', '2ye2', '2ye3', '2ye4', '2ye5', '2ye6', '2ye7', '2ye8', '2ye9', '2yea', '2yeb', '2yec', '2yed', '2yee', '2yef', '2yeg', '2yeh', '2yei', '2yej', '2yi0', '2yi5', '2yi6', '2yi7', '2yjw', '2yjx', '2yk2', '2yk9', '2ykb', '2ykc', '2yke', '2yki', '2ykj', '3b24', '3b25', '3b26', '3b27', '3b28', '3bm9', '3bmy', '3d0b', '3eko', '3ekr', '3ft5', '3ft8', '3hek', '3hhu', '3hyy', '3hyz', '3hz1', '3hz5', '3inw', '3inx', '3k97', '3k98', '3k99', '3mnr', '3o0i', '3ow6', '3owb', '3owd', '3q6m', '3q6n', '3qdd', '3qtf', '3r4m', '3r4n', '3r4o', '3r4p', '3r91', '3r92', '3rkz', '3rlp', '3rlq', '3rlr', '3t0h', '3t0z', '3t10', '3t1k', '3t2s', '3tuh', '3vha', '3vhc', '3vhd', '3wha', '3wq9', '4aif', '4awo', '4awp', '4awq', '4b7p', '4bqg', '4bqj', '4cgq', '4cgu', '4cgv', '4cgw', '4cwf', '4cwn', '4cwo', '4cwp', '4cwq', '4cwr', '4cws', '4cwt', '4eeh', '4eft', '4efu', '4egh', '4egi', '4egk', '4fcp', '4fcq', '4fcr', '4hy6', '4jql', '4l8z', '4l90', '4l91', '4l93', '4l94', '4lwe', '4lwf', '4lwg', '4lwh', '4lwi', '4nh7', '4nh8', '4o04', '4o05', '4o07', '4o09', '4o0b', '4r3m', '4u93', '4w7t', '4xip', '4xiq', '4xir', '4xit', '4ykq', '4ykr', '4ykt', '4yku', '4ykw', '4ykx', '4yky', '4ykz', '5cf0', '5fnc', '5fnd', '5fnf', '5ggz', '5j20', '5j27', '5j2v', '5j2x', '5j64', '5j6l', '5j6m', '5j6n', '5j80', '5j82', '5j86', '5j8m', '5j8u', '5j9x', '5lny', '5lnz', '5lo0', '5lo1', '5lo5', '5lo6', '5lq9', '5lr1', '5lr7', '5lrl', '5lrz', '5ls1', '5m4e', '5m4h', '5nyh', '5nyi', '5oci', '5od7', '5odx', '5t21', '5vyy', '5xqd', '5xqe', '5xr5', '5xr9', '5xrb', '5xrd', '5xre', '5zr3', '6b99', '6b9a', '6ceo', '6cyg', '6cyh', '6ei5', '6el5', '6eln', '6elo', '6elp', '6ey8', '6ey9', '6eya', '6eyb', '6f1n', '6fcj', '6fdp', '6gp4', '6gp8', '6gpf', '6gph', '6gpo', '6gpp', '6gpr', '6gpt', '6gpw', '6gpy', '6gq6', '6gqr', '6gqs', '6gqu', '6gr1', '6gr3', '6gr4', '6gr5', '6hhr', '6ksq', '6lr9', '6lsz', '6lt8', '6lti', '6ltk', '6n8x', '6olx', '6tn4', '6tn5', '6u98', '6u99', '6u9a', '6u9b', '7dmc', '7krj', '7kw7', '7l7i', '7l7j', '7lsz', '7lt0', '7ry1', '7s8y', '7s8z', '7s90', '7s95', '7s98', '7s99', '7s9f', '7s9g', '7s9h', '7s9i', '8agi', '8agj', '8agl']</code></pre>
</div>
</div>
<p>This results in a list of 328 structures. As uniprot isnâ€™t the only source to get that info, letâ€™s check what comes out of the same type of query against the RCSB PDB, should be the same, right?</p>
<div class="cell" data-execution_count="2">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>inputjson<span class="op">=</span><span class="st">"""{</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="st">  "query": {</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="st">    "type": "group",</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="st">    "nodes": [</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="st">      {</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="st">        "type": "terminal",</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="st">        "service": "text",</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="st">        "parameters": {</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="st">          "attribute": "rcsb_polymer_entity_container_identifiers.reference_sequence_identifiers.database_accession",</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="st">          "operator": "in",</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="st">          "value": [</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="st">            "P07900"</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="st">          ]</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="st">        }</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="st">      },</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="st">      {</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="st">        "type": "terminal",</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="st">        "service": "text",</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="st">        "parameters": {</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a><span class="st">          "attribute": "rcsb_polymer_entity_container_identifiers.reference_sequence_identifiers.database_name",</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a><span class="st">          "operator": "exact_match",</span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a><span class="st">          "value": "UniProt",</span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a><span class="st">          "negation": false</span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a><span class="st">        }</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a><span class="st">      }</span></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a><span class="st">    ],</span></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a><span class="st">    "logical_operator": "and",</span></span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a><span class="st">    "label": "nested-attribute"</span></span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a><span class="st">  },</span></span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a><span class="st">  "return_type": "entry",</span></span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a><span class="st">  "request_options": {</span></span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a><span class="st">    "paginate": {</span></span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a><span class="st">      "start": 0,</span></span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a><span class="st">      "rows": 500</span></span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a><span class="st">    },</span></span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a><span class="st">    "results_content_type": [</span></span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a><span class="st">      "experimental"</span></span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a><span class="st">    ],</span></span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a><span class="st">    "sort": [</span></span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a><span class="st">      {</span></span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a><span class="st">        "sort_by": "score",</span></span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a><span class="st">        "direction": "desc"</span></span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a><span class="st">      }</span></span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a><span class="st">    ],</span></span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a><span class="st">    "scoring_strategy": "combined"</span></span>
<span id="cb10-46"><a href="#cb10-46" aria-hidden="true" tabindex="-1"></a><span class="st">  }</span></span>
<span id="cb10-47"><a href="#cb10-47" aria-hidden="true" tabindex="-1"></a><span class="st">}"""</span></span>
<span id="cb10-48"><a href="#cb10-48" aria-hidden="true" tabindex="-1"></a>url<span class="op">=</span><span class="st">"https://search.rcsb.org/rcsbsearch/v2/query?json="</span><span class="op">+</span>inputjson</span>
<span id="cb10-49"><a href="#cb10-49" aria-hidden="true" tabindex="-1"></a>response<span class="op">=</span>requests.get(url)</span>
<span id="cb10-50"><a href="#cb10-50" aria-hidden="true" tabindex="-1"></a>tmp<span class="op">=</span>json.loads(response.text)[<span class="st">"result_set"</span>]</span>
<span id="cb10-51"><a href="#cb10-51" aria-hidden="true" tabindex="-1"></a>rcsb_pdbcodes<span class="op">=</span>[element[<span class="st">"identifier"</span>].lower() <span class="cf">for</span> element <span class="kw">in</span> tmp]</span>
<span id="cb10-52"><a href="#cb10-52" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(rcsb_pdbcodes))</span>
<span id="cb10-53"><a href="#cb10-53" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(rcsb_pdbcodes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>331
['1byq', '1osf', '1uy6', '1uy7', '1uy8', '1uy9', '1uyc', '1uyd', '1uye', '1uyf', '1uyg', '1uyh', '1uyi', '1uyk', '1uyl', '1yc1', '1yc3', '1yc4', '1yer', '1yes', '1yet', '2bsm', '2bt0', '2bug', '2byh', '2byi', '2bz5', '2c2l', '2ccs', '2cct', '2ccu', '2fwy', '2fwz', '2h55', '2jjc', '2k5b', '2qf6', '2qfo', '2qg0', '2qg2', '2uwd', '2vci', '2vcj', '2wi1', '2wi2', '2wi3', '2wi4', '2wi5', '2wi6', '2wi7', '2xab', '2xdk', '2xdl', '2xds', '2xdu', '2xdx', '2xhr', '2xht', '2xhx', '2xjg', '2xjj', '2xjx', '2xk2', '2ye2', '2ye3', '2ye4', '2ye5', '2ye6', '2ye7', '2ye8', '2ye9', '2yea', '2yeb', '2yec', '2yed', '2yee', '2yef', '2yeg', '2yeh', '2yei', '2yej', '2yi0', '2yi5', '2yi6', '2yi7', '2yjw', '2yjx', '2yk2', '2yk9', '2ykb', '2ykc', '2yke', '2yki', '2ykj', '3b24', '3b25', '3b26', '3b27', '3b28', '3bm9', '3bmy', '3d0b', '3eko', '3ekr', '3ft5', '3ft8', '3hek', '3hhu', '3hyy', '3hyz', '3hz1', '3hz5', '3inw', '3inx', '3k97', '3k98', '3k99', '3mnr', '3o0i', '3ow6', '3owb', '3owd', '3q6m', '3q6n', '3qdd', '3qtf', '3r4m', '3r4n', '3r4o', '3r4p', '3r91', '3r92', '3rkz', '3rlp', '3rlq', '3rlr', '3t0h', '3t0z', '3t10', '3t1k', '3t2s', '3tuh', '3vha', '3vhc', '3vhd', '3wha', '3wq9', '4aif', '4awo', '4awp', '4awq', '4b7p', '4bqg', '4bqj', '4cgq', '4cgu', '4cgv', '4cgw', '4cwf', '4cwn', '4cwo', '4cwp', '4cwq', '4cwr', '4cws', '4cwt', '4eeh', '4eft', '4efu', '4egh', '4egi', '4egk', '4fcp', '4fcq', '4fcr', '4hy6', '4jql', '4l8z', '4l90', '4l91', '4l93', '4l94', '4lwe', '4lwf', '4lwg', '4lwh', '4lwi', '4nh7', '4nh8', '4o04', '4o05', '4o07', '4o09', '4o0b', '4r3m', '4u93', '4w7t', '4xip', '4xiq', '4xir', '4xit', '4ykq', '4ykr', '4ykt', '4yku', '4ykw', '4ykx', '4yky', '4ykz', '5cf0', '5fnc', '5fnd', '5fnf', '5ggz', '5j20', '5j27', '5j2v', '5j2x', '5j64', '5j6l', '5j6m', '5j6n', '5j80', '5j82', '5j86', '5j8m', '5j8u', '5j9x', '5lny', '5lnz', '5lo0', '5lo1', '5lo5', '5lo6', '5lq9', '5lr1', '5lr7', '5lrl', '5lrz', '5ls1', '5m4e', '5m4h', '5nyh', '5nyi', '5oci', '5od7', '5odx', '5t21', '5vyy', '5xqd', '5xqe', '5xr5', '5xr9', '5xrb', '5xrd', '5xre', '5zr3', '6b99', '6b9a', '6ceo', '6cyg', '6cyh', '6ei5', '6el5', '6eln', '6elo', '6elp', '6ey8', '6ey9', '6eya', '6eyb', '6f1n', '6fcj', '6fdp', '6gp4', '6gp8', '6gpf', '6gph', '6gpo', '6gpp', '6gpr', '6gpt', '6gpw', '6gpy', '6gq6', '6gqr', '6gqs', '6gqu', '6gr1', '6gr3', '6gr4', '6gr5', '6hhr', '6ksq', '6lr9', '6lsz', '6lt8', '6lti', '6ltk', '6n8x', '6olx', '6tn4', '6tn5', '6u98', '6u99', '6u9a', '6u9b', '7dmc', '7krj', '7kw7', '7l7i', '7l7j', '7lsz', '7lt0', '7rxz', '7ry0', '7ry1', '7s8y', '7s8z', '7s90', '7s95', '7s98', '7s99', '7s9f', '7s9g', '7s9h', '7s9i', '7ur3', '8agi', '8agj', '8agl']</code></pre>
</div>
</div>
<p>From the RCSB we get 3 more structures than from Uniprot. Letâ€™s see what the difference between both lists of pdb codes is:</p>
<p>Again a bit worrysome difference between the RCSB &amp; uniprot. An NCBI search seems to give yet another number of 273 structures. NB: when you have a look at the P07900 sequence cluster page of the RCSB, one gets 332 hits instead. So the RCSB seems to handle the uniprot accession code a bit differently from the Polymer Entity Group ID (or thereâ€™s an issue with the data of one structure).</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Things to keep in mind</p>
<ul>
<li>Uniprot accession vs polymer entity group ID handling to be clarified with RCSB folks</li>
</ul>
</div></div><p>As we can see in <a href="#tbl-atp-binding-site-residues">Table&nbsp;1</a> the residues encompassing our binding site of interest range from amino acid E47 on the WT sequence to V186. So now we need to know whether the region expressed &amp; resolved includes the region of interest. This can now be done with a graphql query versus the RCSB (after a lot of trial &amp; error <span class="emoji" data-emoji="unamused">ðŸ˜’</span> ). NB: I directly filter out structures (18) that contain mutations versus wild type.</p>
<div class="cell" data-execution_count="4">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>query<span class="op">=</span><span class="st">"""</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="st">{</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="st">  entries(entry_ids: """</span><span class="op">+</span><span class="bu">str</span>(rcsb_pdbcodes).replace(<span class="st">"'"</span>,<span class="st">"</span><span class="ch">\"</span><span class="st">"</span>)<span class="op">+</span><span class="st">""")</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="st">  {</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="st">    rcsb_id</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="st">    </span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="st">    polymer_entities {</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="st">      rcsb_polymer_entity_container_identifiers {</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="st">        entity_id</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="st">        asym_ids</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="st">        auth_asym_ids</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="st">        reference_sequence_identifiers {</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="st">          database_name</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="st">          database_accession</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="st">        }</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a><span class="st">      }</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a><span class="st">      rcsb_polymer_entity_align{</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a><span class="st">        </span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a><span class="st">  reference_database_name</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a><span class="st">        aligned_regions{</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a><span class="st">          entity_beg_seq_id</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a><span class="st">          length</span></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a><span class="st">          ref_beg_seq_id</span></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a><span class="st">        }</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a><span class="st">        </span></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a><span class="st">      }</span></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a><span class="st">      entity_poly {</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a><span class="st">        pdbx_seq_one_letter_code_can</span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a><span class="st">        rcsb_sample_sequence_length</span></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a><span class="st">        type</span></span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a><span class="st">        rcsb_mutation_count</span></span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a><span class="st">      }</span></span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a><span class="st">      rcsb_polymer_entity {</span></span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a><span class="st">        formula_weight</span></span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true" tabindex="-1"></a><span class="st">      }</span></span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true" tabindex="-1"></a><span class="st">    }</span></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true" tabindex="-1"></a><span class="st">  }</span></span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true" tabindex="-1"></a><span class="co"># ugly I know, but does the job here</span></span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> extractSequenceRegions(data):</span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true" tabindex="-1"></a>  result<span class="op">=</span>{}</span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> entry <span class="kw">in</span> data:  </span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true" tabindex="-1"></a>    result[entry[<span class="st">"rcsb_id"</span>]]<span class="op">=</span>[]</span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> entity <span class="kw">in</span> entry[<span class="st">"polymer_entities"</span>]:</span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true" tabindex="-1"></a>      <span class="co"># print(entity["rcsb_polymer_entity_container_identifiers"])</span></span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> entity[<span class="st">"entity_poly"</span>][<span class="st">"rcsb_mutation_count"</span>]<span class="op">==</span><span class="dv">0</span>:</span>
<span id="cb12-50"><a href="#cb12-50" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> idx,identifier <span class="kw">in</span> <span class="bu">enumerate</span>(entity[<span class="st">"rcsb_polymer_entity_container_identifiers"</span>][<span class="st">"reference_sequence_identifiers"</span>]):</span>
<span id="cb12-51"><a href="#cb12-51" aria-hidden="true" tabindex="-1"></a>          <span class="cf">if</span> identifier[<span class="st">"database_name"</span>]<span class="op">==</span><span class="st">"UniProt"</span> <span class="kw">and</span> identifier[<span class="st">"database_accession"</span>]<span class="op">==</span><span class="st">"P07900"</span>:</span>
<span id="cb12-52"><a href="#cb12-52" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> entity[<span class="st">"rcsb_polymer_entity_align"</span>][idx][<span class="st">"reference_database_name"</span>]<span class="op">==</span><span class="st">"UniProt"</span>:</span>
<span id="cb12-53"><a href="#cb12-53" aria-hidden="true" tabindex="-1"></a>              <span class="cf">for</span> region <span class="kw">in</span> entity[<span class="st">"rcsb_polymer_entity_align"</span>][idx][<span class="st">"aligned_regions"</span>]:</span>
<span id="cb12-54"><a href="#cb12-54" aria-hidden="true" tabindex="-1"></a>                result[entry[<span class="st">"rcsb_id"</span>]].append((region[<span class="st">"ref_beg_seq_id"</span>],region[<span class="st">"ref_beg_seq_id"</span>]<span class="op">+</span>region[<span class="st">"length"</span>]))</span>
<span id="cb12-55"><a href="#cb12-55" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> result</span>
<span id="cb12-56"><a href="#cb12-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-57"><a href="#cb12-57" aria-hidden="true" tabindex="-1"></a>residuelist<span class="op">=</span>[<span class="dv">47</span>,<span class="dv">48</span>,<span class="dv">51</span>,<span class="dv">52</span>,<span class="dv">54</span>,<span class="dv">55</span>,<span class="dv">58</span>,<span class="dv">93</span>,<span class="dv">95</span>,<span class="dv">96</span>,<span class="dv">97</span>,<span class="dv">98</span>,<span class="dv">106</span>,<span class="dv">107</span>,<span class="dv">112</span>,<span class="dv">131</span>,<span class="dv">132</span>,<span class="dv">133</span>,<span class="dv">134</span>,<span class="dv">135</span>,<span class="dv">136</span>,<span class="dv">137</span>,<span class="dv">138</span>,<span class="dv">139</span>,<span class="dv">152</span>,<span class="dv">183</span>,<span class="dv">184</span>,<span class="dv">186</span>]</span>
<span id="cb12-58"><a href="#cb12-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-59"><a href="#cb12-59" aria-hidden="true" tabindex="-1"></a>url<span class="op">=</span><span class="ss">f"https://data.rcsb.org/graphql?query=</span><span class="sc">{</span>query<span class="sc">}</span><span class="ss">"</span></span>
<span id="cb12-60"><a href="#cb12-60" aria-hidden="true" tabindex="-1"></a>response<span class="op">=</span>requests.get(url)</span>
<span id="cb12-61"><a href="#cb12-61" aria-hidden="true" tabindex="-1"></a>dataRCSBGQL<span class="op">=</span>response.json()[<span class="st">"data"</span>][<span class="st">"entries"</span>]</span>
<span id="cb12-62"><a href="#cb12-62" aria-hidden="true" tabindex="-1"></a>resolved_regions<span class="op">=</span>extractSequenceRegions(dataRCSBGQL)</span>
<span id="cb12-63"><a href="#cb12-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-64"><a href="#cb12-64" aria-hidden="true" tabindex="-1"></a>selectedPdbCodes<span class="op">=</span>[]</span>
<span id="cb12-65"><a href="#cb12-65" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pdbCode <span class="kw">in</span> resolved_regions:</span>
<span id="cb12-66"><a href="#cb12-66" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> mapping <span class="kw">in</span> resolved_regions[pdbCode]:</span>
<span id="cb12-67"><a href="#cb12-67" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="bu">all</span>(i<span class="op">&gt;=</span>mapping[<span class="dv">0</span>] <span class="kw">and</span> i<span class="op">&lt;=</span>mapping[<span class="dv">1</span>] <span class="cf">for</span> i <span class="kw">in</span> residuelist)):</span>
<span id="cb12-68"><a href="#cb12-68" aria-hidden="true" tabindex="-1"></a>      selectedPdbCodes.append(pdbCode)</span>
<span id="cb12-69"><a href="#cb12-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-70"><a href="#cb12-70" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(selectedPdbCodes)</span>
<span id="cb12-71"><a href="#cb12-71" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">str</span>(<span class="bu">len</span>(selectedPdbCodes)) <span class="op">+</span> <span class="st">" structures identified"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>['1BYQ', '1OSF', '1UY6', '1UY7', '1UY8', '1UY9', '1UYC', '1UYD', '1UYE', '1UYF', '1UYG', '1UYH', '1UYI', '1UYK', '1UYL', '1YC1', '1YC3', '1YC4', '1YER', '1YES', '1YET', '2BSM', '2BT0', '2BYH', '2BYI', '2BZ5', '2CCS', '2CCT', '2CCU', '2FWY', '2FWZ', '2H55', '2JJC', '2K5B', '2QF6', '2QFO', '2QG0', '2QG2', '2UWD', '2VCI', '2VCJ', '2WI1', '2WI2', '2WI3', '2WI4', '2WI5', '2WI6', '2WI7', '2XAB', '2XDK', '2XDL', '2XDS', '2XDU', '2XDX', '2XHR', '2XHT', '2XHX', '2XJG', '2XJJ', '2XJX', '2XK2', '2YE2', '2YE3', '2YE4', '2YE5', '2YE6', '2YE7', '2YE8', '2YE9', '2YEA', '2YEB', '2YEC', '2YED', '2YEE', '2YEF', '2YEG', '2YEH', '2YEI', '2YEJ', '2YI0', '2YI5', '2YI6', '2YI7', '2YJW', '2YJX', '2YK2', '2YK9', '2YKB', '2YKC', '2YKE', '2YKI', '2YKJ', '3B24', '3B25', '3B26', '3B27', '3B28', '3BM9', '3BMY', '3D0B', '3EKO', '3EKR', '3FT5', '3FT8', '3HEK', '3HHU', '3HYY', '3HYZ', '3HZ1', '3HZ5', '3INW', '3INX', '3K97', '3K98', '3K99', '3MNR', '3O0I', '3OW6', '3OWB', '3OWD', '3QDD', '3QTF', '3R4M', '3R4N', '3R4O', '3R4P', '3R91', '3R92', '3RKZ', '3RLP', '3RLQ', '3RLR', '3T0H', '3T0Z', '3T10', '3T1K', '3T2S', '3TUH', '3VHA', '3VHC', '3VHD', '3WHA', '3WQ9', '4AWO', '4AWP', '4AWQ', '4B7P', '4BQG', '4BQJ', '4CWF', '4CWN', '4CWO', '4CWP', '4CWQ', '4CWR', '4CWS', '4CWT', '4EEH', '4EFT', '4EFU', '4EGH', '4EGI', '4EGK', '4FCP', '4FCQ', '4FCR', '4HY6', '4JQL', '4L8Z', '4L90', '4L91', '4L93', '4L94', '4LWE', '4LWF', '4LWG', '4LWH', '4LWI', '4NH7', '4NH8', '4O04', '4O05', '4O07', '4O09', '4O0B', '4R3M', '4U93', '4W7T', '4XIP', '4XIQ', '4XIR', '4XIT', '4YKQ', '4YKR', '4YKT', '4YKU', '4YKW', '4YKX', '4YKY', '4YKZ', '5CF0', '5FNC', '5FND', '5FNF', '5GGZ', '5J20', '5J27', '5J2V', '5J2X', '5J64', '5J6L', '5J6M', '5J6N', '5J82', '5J86', '5J8M', '5J8U', '5J9X', '5LNY', '5LNZ', '5LO0', '5LO1', '5LO5', '5LO6', '5LQ9', '5LR1', '5LR7', '5LRL', '5LRZ', '5LS1', '5M4E', '5M4H', '5NYH', '5NYI', '5OCI', '5OD7', '5ODX', '5T21', '5VYY', '5XQD', '5XQE', '5XR5', '5XR9', '5XRB', '5XRD', '5XRE', '5ZR3', '6B99', '6B9A', '6CEO', '6CYG', '6CYH', '6EI5', '6EL5', '6ELN', '6ELO', '6ELP', '6EY8', '6EY9', '6EYA', '6EYB', '6F1N', '6FCJ', '6GPR', '6GPT', '6GPW', '6GPY', '6GQ6', '6GQR', '6HHR', '6LR9', '6LSZ', '6LT8', '6LTI', '6LTK', '6N8X', '6TN4', '6TN5', '6U99', '6U9B', '7DMC', '7KRJ', '7KW7', '7L7I', '7L7J', '7LSZ', '7LT0', '7S8Y', '7S8Z', '7S90', '7S95', '7S98', '7S99', '7S9F', '7S9G', '7S9H', '7S9I', '7UR3', '8AGI']
299 structures identified</code></pre>
</div>
</div>
<p>Now we have our list of 299 WT structures with the ATP binding site resolved. NB: you might have noticed that Iâ€™m not taking the resolution, Rfree or Rwork into account for the structure selection here. During the subsequent clustering step Iâ€™ll still check if all residues of interest of the binding site are well resolved (some might not be).</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Things to keep in mind</p>
<ul>
<li>Uniprot accession vs polymer entity group ID handling to be clarified with RCSB folks</li>
<li>Donâ€™t forget to check if all sidechains of the binding site are resolved in the selected structures</li>
</ul>
</div></div></section>
<section id="cluster-by-conformation" class="level3 page-columns page-full">
<h3 class="anchored" data-anchor-id="cluster-by-conformation">Cluster by conformation</h3>
<p>The ATP binding site of HSP90 can be fairly flexible in specific regions. In order to evaluate a pocket comparison methodâ€™s sensitivity to sidechain or backbone movements we can split down the 299 structures we just identified to smaller clusters of very similar conformations. This will allow us later to run tests of intra vs extra cluster pocket comparison using different algorithms.</p>
<p>Already doing all the previous work was not as straightforward what one might have expected, but itâ€™s still doable. Especially with the API available now from the RCSB. A few years back one had to rely on predictions from the <a href="https://www.ebi.ac.uk/pdbe/docs/sifts/">SIFTS project</a> from <span class="citation" data-cites="Dana2018">Dana et al. (<a href="#ref-Dana2018" role="doc-biblioref">2018</a>)</span> &amp; <span class="citation" data-cites="Velankar2012">Velankar et al. (<a href="#ref-Velankar2012" role="doc-biblioref">2012</a>)</span>.</p>
<p>Structural superimpositioning poses the following challenges here:</p>
<ol type="1">
<li>shall we use all chains or all if the asymetric unit contains several copies of the same protein?
<ul>
<li>In theory we should assess if there are major differences between monomers or use the monomers as expected hit we should retrieve when comparing binding sites. Both should be covered by the overall dataset here so Iâ€™ll be lazy and will focus on the first chain</li>
</ul></li>
<li>shall we use a standard Kabsch superimpositioning (local/global?) &amp; overall root mean square deviation to assess the similarity between binding site conformations?
<ul>
<li>If Iâ€™m bringing this thing up, I think itâ€™s important to address it properly. Iâ€™ll try to explain a bit in more detail why.</li>
</ul></li>
</ol>

<div class="no-row-height column-margin column-container"><div class="">
<p>Things to keep in mind</p>
<ul>
<li>Uniprot accession vs polymer entity group ID handling to be clarified with RCSB folks</li>
<li>Donâ€™t forget to check if all sidechains of the binding site are resolved in the selected structures</li>
<li>Treat homo-multimers rationally in the data-set construction</li>
</ul>
</div></div><section id="why-isnt-standard-superimpositioning-enough" class="level4">
<h4 class="anchored" data-anchor-id="why-isnt-standard-superimpositioning-enough">Why isnâ€™t standard superimpositioning enough?</h4>
<p>I donâ€™t have to write a ton of text here, citing the single, most relevant paper from <span class="citation" data-cites="Damm2006">Damm and Carlson (<a href="#ref-Damm2006" role="doc-biblioref">2006</a>)</span> should be argument enough <span class="emoji" data-emoji="grin">ðŸ˜</span></p>
<p>But to showcase maybe a bit more clearly. HSP90 can adopt several conformations in different areas of the binding site:</p>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb14" data-startfrom="514" data-source-offset="0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 513;"><span id="cb14-514"><a href="#cb14-514" aria-hidden="true" tabindex="-1"></a><span class="co">// Create drawing area</span></span>
<span id="cb14-515"><a href="#cb14-515" aria-hidden="true" tabindex="-1"></a>divNGL2 <span class="op">=</span> <span class="fu">html</span><span class="vs">`&lt;div style="width:800px;height:600px;position:relative"&gt;&lt;/div&gt;`</span><span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-hsp-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div id="ojs-cell-3" data-nodetype="declaration">

</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;2: Comparison of distinct binding site conformations (3t0z,4cwr &amp; 7l7j)</figcaption><p></p>
</figure>
</div>
</div>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb15" data-startfrom="523" data-source-offset="-0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 522;"><span id="cb15-523"><a href="#cb15-523" aria-hidden="true" tabindex="-1"></a>stage2 <span class="op">=</span> <span class="kw">new</span> NGL<span class="op">.</span><span class="fu">Stage</span>(divNGL2<span class="op">,</span> { <span class="dt">backgroundColor</span><span class="op">:</span> <span class="st">"black"</span><span class="op">,</span><span class="dt">cameraType</span><span class="op">:</span> <span class="st">"orthographic"</span>})<span class="op">;</span></span>
<span id="cb15-524"><a href="#cb15-524" aria-hidden="true" tabindex="-1"></a>pdbString1 <span class="op">=</span> <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"3t0z.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">;</span></span>
<span id="cb15-525"><a href="#cb15-525" aria-hidden="true" tabindex="-1"></a>structure1 <span class="op">=</span> <span class="cf">await</span> stage2<span class="op">.</span><span class="fu">loadFile</span>(pdbString1<span class="op">,</span> {<span class="dt">ext</span><span class="op">:</span> <span class="st">"pdb"</span><span class="op">,</span> <span class="dt">asTrajectory</span><span class="op">:</span> <span class="kw">false</span>})</span>
<span id="cb15-526"><a href="#cb15-526" aria-hidden="true" tabindex="-1"></a>structure1<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"cartoon"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"protein"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#325E73"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-2" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-3" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-4" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb16" data-startfrom="527" data-source-offset="-307"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 526;"><span id="cb16-527"><a href="#cb16-527" aria-hidden="true" tabindex="-1"></a>structure1<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"ball+stick"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"ATP"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"element"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-5" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb17" data-startfrom="528" data-source-offset="-388"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 527;"><span id="cb17-528"><a href="#cb17-528" aria-hidden="true" tabindex="-1"></a>pdbString2 <span class="op">=</span> <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/4cwr_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">;</span></span>
<span id="cb17-529"><a href="#cb17-529" aria-hidden="true" tabindex="-1"></a>structure2 <span class="op">=</span> <span class="cf">await</span> stage2<span class="op">.</span><span class="fu">loadFile</span>(pdbString2<span class="op">,</span> {<span class="dt">ext</span><span class="op">:</span> <span class="st">"pdb"</span><span class="op">,</span> <span class="dt">asTrajectory</span><span class="op">:</span> <span class="kw">false</span>})</span>
<span id="cb17-530"><a href="#cb17-530" aria-hidden="true" tabindex="-1"></a>structure2<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"cartoon"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"protein"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#B69849"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-6" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-7" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-8" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb18" data-startfrom="531" data-source-offset="-619"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 530;"><span id="cb18-531"><a href="#cb18-531" aria-hidden="true" tabindex="-1"></a>pdbString3 <span class="op">=</span> <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/7l7j_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">;</span></span>
<span id="cb18-532"><a href="#cb18-532" aria-hidden="true" tabindex="-1"></a>structure3 <span class="op">=</span> <span class="cf">await</span> stage2<span class="op">.</span><span class="fu">loadFile</span>(pdbString3<span class="op">,</span> {<span class="dt">ext</span><span class="op">:</span> <span class="st">"pdb"</span><span class="op">,</span> <span class="dt">asTrajectory</span><span class="op">:</span> <span class="kw">false</span>})</span>
<span id="cb18-533"><a href="#cb18-533" aria-hidden="true" tabindex="-1"></a>structure3<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"cartoon"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"protein"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#B66049"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-9" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-10" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-11" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb19" data-startfrom="535" data-source-offset="-851"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 534;"><span id="cb19-535"><a href="#cb19-535" aria-hidden="true" tabindex="-1"></a>structure1<span class="op">.</span><span class="fu">autoView</span>(<span class="st">"ATP"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-4-12" data-nodetype="expression">

</div>
</div>
</div>
</div>
<p>So if you superimpose binding site residues only with a classical Kabsch algorithm (<span class="citation" data-cites="Kabsch1976">Kabsch (<a href="#ref-Kabsch1976" role="doc-biblioref">1976</a>)</span>) you might alter significantly the positions of all binding site residues that appear to be stable (beta sheet &amp; helix 44-58 for instance). Also clustering such superposed structures on differences on amino acid positions will yield sub-optimal results.</p>
<p>Furthermore, one should decouple the act of superimposing two structures &amp; the act of calculating the similarity of conformations once they are superimposed. With the classical Kabsch algorithm &amp; the resulting RMSD, all of this is intertwined.</p>
<p><span class="citation" data-cites="Kufareva2011">Kufareva and Abagyan (<a href="#ref-Kufareva2011" role="doc-biblioref">2011</a>)</span> already wrote about that topic and stated the properties of an ideal protein similarity measure as:</p>
<blockquote class="blockquote">
<p>An ideal measure should allow both a single â€˜summaryâ€™ number within a fixed range (e.g.&nbsp;0% to 100%) and an underlying detailed vector or matrix representation. The single number must distinguish well between related (correct) and non-related (incorrect) structure pairs, i.e.&nbsp;its distributions on the two sets must overlap to a minimal possible degree. It has to be relevant, i.e.&nbsp;capture the nature of protein folding or protein interaction determinants rather than satisfy simple geometric criteria. It has to have the minimal number of parameters, which in turn need to be well justified and understandable. It has to be stable and robust against minor or fractional (affecting a small fraction of the model) experimental and modeling errors; such changes in the structures should not lead to major leaps in the calculated similarity measure values. It has to capture the similarities or differences between the structures at any given level of accuracy/resolution. Ideally, it should have an intuitive visual interpretation.</p>
</blockquote>
<p><span class="citation" data-cites="Kufareva2011">Kufareva and Abagyan (<a href="#ref-Kufareva2011" role="doc-biblioref">2011</a>)</span> however conclude that even distance weighted optimization &amp; related cartesian space metrics are always prone to several issues. In the end they deem that contact based metrics are more robust. So this is what Iâ€™ll be using here.</p>
</section>
<section id="contact-based-conformation-comparison" class="level4 page-columns page-full">
<h4 class="anchored" data-anchor-id="contact-based-conformation-comparison">Contact based conformation comparison</h4>
<p>Here Iâ€™m going to use some tools already existing in MDanalysis, a popular python toolkit for MD trajectory analysis. It contains already prebuilt funtions to calculate the contact maps of a protein structure. But overall, you can achieve the same thing with any other toolkit &amp; a lot of looping.</p>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> MDAnalysis <span class="im">as</span> mda</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.spatial.distance <span class="im">import</span> cdist</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> MDAnalysis.analysis <span class="im">import</span> distances</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> gemmi</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>filename<span class="op">=</span><span class="st">"3t0z.pdb"</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>u <span class="op">=</span> mda.Universe(filename)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>residue_seqnums<span class="op">=</span>[<span class="dv">47</span>,<span class="dv">48</span>,<span class="dv">51</span>,<span class="dv">52</span>,<span class="dv">54</span>,<span class="dv">55</span>,<span class="dv">58</span>,<span class="dv">93</span>,<span class="dv">95</span>,<span class="dv">96</span>,<span class="dv">97</span>,<span class="dv">98</span>,<span class="dv">106</span>,<span class="dv">107</span>,<span class="dv">112</span>,<span class="dv">131</span>,<span class="dv">132</span>,<span class="dv">133</span>,<span class="dv">134</span>,<span class="dv">135</span>,<span class="dv">136</span>,<span class="dv">137</span>,<span class="dv">138</span>,<span class="dv">139</span>,<span class="dv">152</span>,<span class="dv">183</span>,<span class="dv">184</span>,<span class="dv">186</span>]</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>residues <span class="op">=</span> u.select_atoms(<span class="st">'protein and name CA and resnum </span><span class="sc">{}</span><span class="st">'</span>.<span class="bu">format</span>(<span class="st">' '</span>.join(<span class="bu">map</span>(<span class="bu">str</span>, residue_seqnums))))</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(residues)</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>distances3t0z <span class="op">=</span> distances.distance_array(residues.positions, residues.positions)</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>sns.heatmap(distances3t0z,yticklabels<span class="op">=</span>residue_seqnums,xticklabels<span class="op">=</span>residue_seqnums,cmap<span class="op">=</span><span class="st">"Blues"</span>,square<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;AtomGroup [&lt;Atom 302: CA of type C of resname GLU, resid 47 and segid A and altLoc &gt;, &lt;Atom 311: CA of type C of resname LEU, resid 48 and segid A and altLoc &gt;, &lt;Atom 333: CA of type C of resname ASN, resid 51 and segid A and altLoc &gt;, ..., &lt;Atom 1331: CA of type C of resname GLY, resid 183 and segid A and altLoc &gt;, &lt;Atom 1335: CA of type C of resname THR, resid 184 and segid A and altLoc &gt;, &lt;Atom 1351: CA of type C of resname VAL, resid 186 and segid A and altLoc &gt;]&gt;</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>&lt;Axes: &gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="2023-04-03-binding-site-comparison-benchmark-I_files/figure-html/cell-6-output-3.png" width="494" height="424"></p>
</div>
</div>
<p>The code above has two main limitations that come to my mind. The first one is that the residue numbering Iâ€™m using is the canonical uniprot sequence one, but Iâ€™m not assured the residue numbering is the same in all HSP90 structures I selected. The second one is that Iâ€™m only computing distances between alpha carbons. This implies that Iâ€™ll only be able to measure differences in backbone conformations and as such important conformational changes. I wonâ€™t be able to detect subtle side-chain conformational changes. One could use an all atom distance matrix, or compute the center of mass for each residue &amp; compare that instead. The time being letâ€™s stick to alpha carbons. Also, I dont need to check for missing atoms in the structures for now with this simplification.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Things to keep in mind</p>
<ul>
<li>Uniprot accession vs polymer entity group ID handling to be clarified with RCSB folks</li>
<li>Donâ€™t forget to check if all sidechains of the binding site are resolved in the selected structures</li>
<li>Treat homo-multimers rationally in the data-set construction</li>
<li>Handle canonical residue numbering properly in distanceMatrix code</li>
<li>Consider also optional side-chain sensitive distance calculations</li>
</ul>
</div></div><p>Now letâ€™s calculate the same distance matrix on another structure, for instance 4cwr.</p>
<div class="cell" data-execution_count="6">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>filename<span class="op">=</span><span class="st">"pdbs/4cwr_aligned.pdb"</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>u <span class="op">=</span> mda.Universe(filename)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>residues <span class="op">=</span> u.select_atoms(<span class="st">'protein and name CA and resnum </span><span class="sc">{}</span><span class="st">'</span>.<span class="bu">format</span>(<span class="st">' '</span>.join(<span class="bu">map</span>(<span class="bu">str</span>, residue_seqnums))))</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>distances4cwr <span class="op">=</span> distances.distance_array(residues.positions, residues.positions)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>,<span class="dv">2</span>)</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>sns.heatmap(distances3t0z,yticklabels<span class="op">=</span>residue_seqnums,xticklabels<span class="op">=</span>residue_seqnums,cmap<span class="op">=</span><span class="st">"Blues"</span>,square<span class="op">=</span><span class="va">True</span>,ax<span class="op">=</span>ax[<span class="dv">0</span>])</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>sns.heatmap(distances4cwr,yticklabels<span class="op">=</span>residue_seqnums,xticklabels<span class="op">=</span>residue_seqnums,cmap<span class="op">=</span><span class="st">"Blues"</span>,square<span class="op">=</span><span class="va">True</span>,ax<span class="op">=</span>ax[<span class="dv">1</span>])</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-dm-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-04-03-binding-site-comparison-benchmark-I_files/figure-html/fig-dm-1-output-1.png" width="584" height="397" class="figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;3: Comparison of distinct binding site conformations (3t0z,4cwr) using a CA distance matrix</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>Visual inspection wonâ€™t help a lot here. Letâ€™s calculate the difference between the two matrices. Iâ€™ll call the resulting matrix a <em>contact difference matrix</em> from hereon. And check if we can spot the residues we observe to be moving significantly.</p>
<div class="cell" data-execution_count="7">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>distancematrix<span class="op">=</span>np.<span class="bu">abs</span>(distances3t0z<span class="op">-</span>distances4cwr)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>sns.heatmap(distancematrix,yticklabels<span class="op">=</span>residue_seqnums,xticklabels<span class="op">=</span>residue_seqnums,cmap<span class="op">=</span><span class="st">"Blues"</span>,square<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-dm-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-04-03-binding-site-comparison-benchmark-I_files/figure-html/fig-dm-2-output-1.png" width="499" height="424" class="figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;4: Comparison of distinct binding site conformations (3t0z,4cwr) using a CA distance matrix</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>That looks already a lot nicer. Here we can see that residues 106, 107 and 112 clearly move a lot, comparing both structures. Also the section 131-139 shows signficant motion when compared to residues 41-98. To understand better what is going on, letâ€™s show both structures overlayed again &amp; highlight &amp; label all implicated residues.</p>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb25" data-startfrom="628" data-source-offset="0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 627;"><span id="cb25-628"><a href="#cb25-628" aria-hidden="true" tabindex="-1"></a><span class="co">// Create drawing area</span></span>
<span id="cb25-629"><a href="#cb25-629" aria-hidden="true" tabindex="-1"></a>divNGL3 <span class="op">=</span> <span class="fu">html</span><span class="vs">`&lt;div style="width:800px;height:600px;position:relative"&gt;&lt;/div&gt;`</span><span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-hsp-3" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div id="ojs-cell-5" data-nodetype="declaration">

</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;5: Comparison of distinct binding site conformations (3t0z,4cwr) on specific residues</figcaption><p></p>
</figure>
</div>
</div>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb26" data-startfrom="637" data-source-offset="-0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 636;"><span id="cb26-637"><a href="#cb26-637" aria-hidden="true" tabindex="-1"></a>stage3 <span class="op">=</span> <span class="kw">new</span> NGL<span class="op">.</span><span class="fu">Stage</span>(divNGL3<span class="op">,</span> { <span class="dt">backgroundColor</span><span class="op">:</span> <span class="st">"black"</span><span class="op">,</span><span class="dt">cameraType</span><span class="op">:</span> <span class="st">"orthographic"</span>})<span class="op">;</span></span>
<span id="cb26-638"><a href="#cb26-638" aria-hidden="true" tabindex="-1"></a>pdbString31 <span class="op">=</span> <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"3t0z.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">;</span></span>
<span id="cb26-639"><a href="#cb26-639" aria-hidden="true" tabindex="-1"></a>structure31 <span class="op">=</span> <span class="cf">await</span> stage3<span class="op">.</span><span class="fu">loadFile</span>(pdbString31<span class="op">,</span> {<span class="dt">ext</span><span class="op">:</span> <span class="st">"pdb"</span><span class="op">,</span> <span class="dt">asTrajectory</span><span class="op">:</span> <span class="kw">false</span>})</span>
<span id="cb26-640"><a href="#cb26-640" aria-hidden="true" tabindex="-1"></a>structure31<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"trace"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"protein"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#325E73"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-2" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-3" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-4" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb27" data-startfrom="641" data-source-offset="-309"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 640;"><span id="cb27-641"><a href="#cb27-641" aria-hidden="true" tabindex="-1"></a>structure31<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"ball+stick"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"(106-107 or 112 or 131-139) and .CA"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#325E73"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-5" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb28" data-startfrom="642" data-source-offset="-423"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 641;"><span id="cb28-642"><a href="#cb28-642" aria-hidden="true" tabindex="-1"></a>structure31<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"label"</span><span class="op">,</span> {<span class="st">"sele"</span><span class="op">:</span> <span class="st">"(106-107 or 112 or 131-139) and .CA"</span><span class="op">,</span> <span class="st">"labelType"</span><span class="op">:</span><span class="st">"resno"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#7F9DAC"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-6" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb29" data-startfrom="644" data-source-offset="-554"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 643;"><span id="cb29-644"><a href="#cb29-644" aria-hidden="true" tabindex="-1"></a>pdbString32 <span class="op">=</span> <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/4cwr_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">;</span></span>
<span id="cb29-645"><a href="#cb29-645" aria-hidden="true" tabindex="-1"></a>structure32 <span class="op">=</span> <span class="cf">await</span> stage3<span class="op">.</span><span class="fu">loadFile</span>(pdbString32<span class="op">,</span> {<span class="dt">ext</span><span class="op">:</span> <span class="st">"pdb"</span><span class="op">,</span> <span class="dt">asTrajectory</span><span class="op">:</span> <span class="kw">false</span>})<span class="op">;</span></span>
<span id="cb29-646"><a href="#cb29-646" aria-hidden="true" tabindex="-1"></a>structure32<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"trace"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"protein"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#B69849"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-7" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-8" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-9" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb30" data-startfrom="647" data-source-offset="-788"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 646;"><span id="cb30-647"><a href="#cb30-647" aria-hidden="true" tabindex="-1"></a>structure32<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"ball+stick"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"(106-107 or 112 or 131-139) and .CA and :A"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#B69849"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-10" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb31" data-startfrom="648" data-source-offset="-909"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 647;"><span id="cb31-648"><a href="#cb31-648" aria-hidden="true" tabindex="-1"></a>structure32<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"label"</span><span class="op">,</span> {<span class="st">"sele"</span><span class="op">:</span> <span class="st">"(106-107 or 112 or 131-139) and .CA and :A"</span><span class="op">,</span> <span class="st">"labelType"</span><span class="op">:</span><span class="st">"resno"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#FFEBB8"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-11" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb32" data-startfrom="649" data-source-offset="-1046"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 648;"><span id="cb32-649"><a href="#cb32-649" aria-hidden="true" tabindex="-1"></a>structure31<span class="op">.</span><span class="fu">autoView</span>(<span class="st">"(106-107 or 112 or 131-139) and .CA and :A"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-6-12" data-nodetype="expression">

</div>
</div>
</div>
</div>
<p>The distance matrix nicely catches the small shift of the 131-139 region. There is a subtle backbone movement ongoing. Regarding residues 106, 107 and 112 they are involved in a helical region that unfolds in one of the two structures. Observing that none of the residues between 107 and 112 are in the distance matrix is due to the fact that we did our initial residue selection based on a single reference structure to define the binding site. In the structure 4cwr this segment however structures itself into a helix. By doing so for instance I11O is now exposed to the binding site lumen. As a result, do we need to adjust the binding site residue selection? In theory, if we want to properly track &amp; understand structural changes in the dataset at hand, yes. In this particular example the difference is captured by the sourrounding residues 106-107 and 112. So letâ€™s put it on my list of things to keep in mind which is getting longer <span class="emoji" data-emoji="scream">ðŸ˜±</span>.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Things to keep in mind</p>
<ul>
<li>Uniprot accession vs polymer entity group ID handling to be clarified with RCSB folks</li>
<li>Donâ€™t forget to check if all sidechains of the binding site are resolved in the selected structures</li>
<li>Treat homo-multimers rationally in the data-set construction</li>
<li>Handle canonical residue numbering properly in distanceMatrix code</li>
<li>Consider also optional side-chain sensitive distance calculations</li>
<li>What to do with large motions implicating new residues in binding sites?</li>
</ul>
</div></div><p>Iâ€™m interested to check on how such a contact difference matrix looks like on very similar structures, so I prepared an example between 4cwr &amp; 1uyd, which a very close (as a lot of the HSP90 alpha human structures actually).</p>
<div class="cell" data-execution_count="8">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>filename<span class="op">=</span><span class="st">"pdbs/1uyd.pdb"</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>u <span class="op">=</span> mda.Universe(filename)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>residues <span class="op">=</span> u.select_atoms(<span class="st">'protein and name CA and resnum </span><span class="sc">{}</span><span class="st">'</span>.<span class="bu">format</span>(<span class="st">' '</span>.join(<span class="bu">map</span>(<span class="bu">str</span>, residue_seqnums))))</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>distances1uyd <span class="op">=</span> distances.distance_array(residues.positions, residues.positions)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>distancematrix<span class="op">=</span>np.<span class="bu">abs</span>(distances4cwr<span class="op">-</span>distances1uyd)</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>sns.heatmap(distancematrix,yticklabels<span class="op">=</span>residue_seqnums,xticklabels<span class="op">=</span>residue_seqnums,cmap<span class="op">=</span><span class="st">"Blues"</span>,square<span class="op">=</span><span class="va">True</span>,vmax<span class="op">=</span><span class="fl">3.7</span>)</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-dm-3" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-04-03-binding-site-comparison-benchmark-I_files/figure-html/fig-dm-3-output-1.png" width="499" height="424" class="figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;6: Comparison of distinct binding site conformations (4cwe,1uyd) using a CA distance matrix</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>I used the same color scale as on <a href="#fig-dm-2">Figure&nbsp;4</a> on purpose, just to have something slightly comparable (visually). However, if you look in more detail, youâ€™ll notice that there are differences (around 0.7A max) between parts of the binding site. But obviously, they are less drastic conformational changes than what we observed before with the comparison of 3t0z &amp; 4cwr.</p>
<p>Last example that I wanted to check is the impact on such a matrix of the closed conformation of the binding site. This is structure 7l7j which was already displayed in red in <a href="#fig-hsp-2">Figure&nbsp;2</a> before.</p>
<div class="cell" data-execution_count="9">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>filename<span class="op">=</span><span class="st">"pdbs/7l7j_aligned.pdb"</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>u <span class="op">=</span> mda.Universe(filename)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>residues <span class="op">=</span> u.select_atoms(<span class="st">'protein and chainID A and name CA and resnum </span><span class="sc">{}</span><span class="st">'</span>.<span class="bu">format</span>(<span class="st">' '</span>.join(<span class="bu">map</span>(<span class="bu">str</span>, residue_seqnums))))</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>distances7l7j <span class="op">=</span> distances.distance_array(residues.positions, residues.positions)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>distancematrix<span class="op">=</span>np.<span class="bu">abs</span>(distances4cwr<span class="op">-</span>distances7l7j)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>sns.heatmap(distancematrix,yticklabels<span class="op">=</span>residue_seqnums,xticklabels<span class="op">=</span>residue_seqnums,cmap<span class="op">=</span><span class="st">"Blues"</span>,square<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-dm-4" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-04-03-binding-site-comparison-benchmark-I_files/figure-html/fig-dm-4-output-1.png" width="486" height="424" class="figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;7: Comparison of distinct binding site conformations (4cwe,7l7j) using a CA distance matrix</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>In the figure above I didnâ€™t set a maximum to the color scale. This way you can see the full extent of movements observed especially around residue 112 - 133. Again, letâ€™s repeat the exercice and plot the residues that appear to be significantly moving here: 106,112, 131-133</p>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb35" data-startfrom="709" data-source-offset="0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 708;"><span id="cb35-709"><a href="#cb35-709" aria-hidden="true" tabindex="-1"></a><span class="co">// Create drawing area</span></span>
<span id="cb35-710"><a href="#cb35-710" aria-hidden="true" tabindex="-1"></a>divNGL4 <span class="op">=</span> <span class="fu">html</span><span class="vs">`&lt;div style="width:800px;height:600px;position:relative"&gt;&lt;/div&gt;`</span><span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-hsp-4" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div id="ojs-cell-7" data-nodetype="declaration">

</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;8: Comparison of distinct binding site conformations (4cwr, 7l7j) on specific residues</figcaption><p></p>
</figure>
</div>
</div>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb36" data-startfrom="718" data-source-offset="-0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 717;"><span id="cb36-718"><a href="#cb36-718" aria-hidden="true" tabindex="-1"></a>stage4 <span class="op">=</span> <span class="kw">new</span> NGL<span class="op">.</span><span class="fu">Stage</span>(divNGL4<span class="op">,</span> { <span class="dt">backgroundColor</span><span class="op">:</span> <span class="st">"black"</span><span class="op">,</span><span class="dt">cameraType</span><span class="op">:</span> <span class="st">"orthographic"</span>})<span class="op">;</span></span>
<span id="cb36-719"><a href="#cb36-719" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-720"><a href="#cb36-720" aria-hidden="true" tabindex="-1"></a>pdbString42 <span class="op">=</span> <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/4cwr_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">;</span></span>
<span id="cb36-721"><a href="#cb36-721" aria-hidden="true" tabindex="-1"></a>structure42 <span class="op">=</span> <span class="cf">await</span> stage4<span class="op">.</span><span class="fu">loadFile</span>(pdbString42<span class="op">,</span> {<span class="dt">ext</span><span class="op">:</span> <span class="st">"pdb"</span><span class="op">,</span> <span class="dt">asTrajectory</span><span class="op">:</span> <span class="kw">false</span>})<span class="op">;</span></span>
<span id="cb36-722"><a href="#cb36-722" aria-hidden="true" tabindex="-1"></a>structure42<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"trace"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"protein"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#B69849"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-2" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-3" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-4" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb37" data-startfrom="723" data-source-offset="-324"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 722;"><span id="cb37-723"><a href="#cb37-723" aria-hidden="true" tabindex="-1"></a>structure42<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"ball+stick"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"(106 or 112 or 131-133) and .CA"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#B69849"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-5" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb38" data-startfrom="724" data-source-offset="-434"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 723;"><span id="cb38-724"><a href="#cb38-724" aria-hidden="true" tabindex="-1"></a>structure42<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"label"</span><span class="op">,</span> {<span class="st">"sele"</span><span class="op">:</span> <span class="st">"(106 or 112 or 131-133) and .CA"</span><span class="op">,</span> <span class="st">"labelType"</span><span class="op">:</span><span class="st">"resno"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#FFEBB8"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-6" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb39" data-startfrom="725" data-source-offset="-560"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 724;"><span id="cb39-725"><a href="#cb39-725" aria-hidden="true" tabindex="-1"></a>pdbString41 <span class="op">=</span> <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/7l7j_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">;</span></span>
<span id="cb39-726"><a href="#cb39-726" aria-hidden="true" tabindex="-1"></a>structure41 <span class="op">=</span> <span class="cf">await</span> stage4<span class="op">.</span><span class="fu">loadFile</span>(pdbString41<span class="op">,</span> {<span class="dt">ext</span><span class="op">:</span> <span class="st">"pdb"</span><span class="op">,</span> <span class="dt">asTrajectory</span><span class="op">:</span> <span class="kw">false</span>})<span class="op">;</span></span>
<span id="cb39-727"><a href="#cb39-727" aria-hidden="true" tabindex="-1"></a>structure41<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"trace"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"protein"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#B66049"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-7" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-8" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-9" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb40" data-startfrom="728" data-source-offset="-794"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 727;"><span id="cb40-728"><a href="#cb40-728" aria-hidden="true" tabindex="-1"></a>structure41<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"ball+stick"</span><span class="op">,</span>{ <span class="st">"sele"</span><span class="op">:</span> <span class="st">"(106 or 112 or 131-133) and .CA and :A"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#B66049"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-10" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb41" data-startfrom="729" data-source-offset="-911"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 728;"><span id="cb41-729"><a href="#cb41-729" aria-hidden="true" tabindex="-1"></a>structure41<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"label"</span><span class="op">,</span> {<span class="st">"sele"</span><span class="op">:</span> <span class="st">"(106 or 112 or 131-133) and .CA and :A"</span><span class="op">,</span> <span class="st">"labelType"</span><span class="op">:</span><span class="st">"resno"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> <span class="st">"#FFC7B8"</span>})<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-11" data-nodetype="expression">

</div>
</div>
</div>
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb42" data-startfrom="730" data-source-offset="-1044"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 729;"><span id="cb42-730"><a href="#cb42-730" aria-hidden="true" tabindex="-1"></a>structure41<span class="op">.</span><span class="fu">autoView</span>(<span class="st">"(106 or 112 or 131-133) and .CA and :A"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-8-12" data-nodetype="expression">

</div>
</div>
</div>
</div>
<p>The conformational changes are very important (as expected). Furthermore the selection of residues one could extract from the distance matrix is relevant also versus what can be observed above.</p>
</section>
<section id="clustering-contact-distance-matrices" class="level4 page-columns page-full">
<h4 class="anchored" data-anchor-id="clustering-contact-distance-matrices">Clustering contact (distance) matrices</h4>
<p>Now, in order to know whether one structure is close to another we can use these contact matrices calculated before &amp; compare these one to another. The comparison between 4cwr &amp; 3t0z results in a contact difference matrix (<a href="#fig-dm-2">Figure&nbsp;4</a>). The comparison of 4cwr and 1uyd results in another contact difference matrix (<a href="#fig-dm-3">Figure&nbsp;6</a>). Letâ€™s try to compute an average of each contact difference matrix and see what we get.</p>
<div class="cell" data-execution_count="10">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>distanceMatrixList<span class="op">=</span>[distances4cwr,distances3t0z,distances1uyd,distances7l7j]</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>labels<span class="op">=</span>[<span class="st">"4cwr"</span>,<span class="st">"3t0z"</span>,<span class="st">"1uyd"</span>,<span class="st">"7l7j"</span>]</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>nMatrices<span class="op">=</span><span class="bu">len</span>(distanceMatrixList)</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>result<span class="op">=</span>np.zeros((nMatrices,nMatrices))</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(nMatrices):</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(nMatrices):</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i<span class="op">==</span>j:</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>      result[i][j]<span class="op">=</span><span class="fl">0.0</span></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> i<span class="op">&lt;</span>j:</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>      result[i][j]<span class="op">=</span>np.mean(np.<span class="bu">abs</span>(distanceMatrixList[i]<span class="op">-</span>distanceMatrixList[j]))</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>      result[j][i]<span class="op">=</span>result[i][j]</span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a>sns.heatmap(result,yticklabels<span class="op">=</span>labels,xticklabels<span class="op">=</span>labels,cmap<span class="op">=</span><span class="st">"Blues"</span>,square<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-dm-pocket-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-04-03-binding-site-comparison-benchmark-I_files/figure-html/fig-dm-pocket-1-output-1.png" width="486" height="411" class="figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;9: Comparison of distinct binding sites for clustering</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>That looks already rather promising. We expect to see 4cwr &amp; 1uyd to be close to each other, and 3t0z to be closer to 4cwr than to 7l7j. However, 7l7j is not close to any of the other structures and likely closest to 3t0z (as both have the helical unfolding around residues 107-112). Now we have a distance matrix that we can feed into a clustering algorithm. Iâ€™m not a big fan of k-means as I donâ€™t know how many clusters I expect beforehand. In the end I want the dataset preparation steps Iâ€™m going through here to be as automated as possible. So Iâ€™ll use a hierarchical clustering algorithm instead.</p>
<div class="cell" data-execution_count="11">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.spatial.distance <span class="im">as</span> ssd</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>distArray <span class="op">=</span> ssd.squareform(result)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>clusters<span class="op">=</span>scipy.cluster.hierarchy.linkage(distArray, method<span class="op">=</span><span class="st">'single'</span>, metric<span class="op">=</span><span class="st">'euclidean'</span>)</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>scipy.cluster.hierarchy.dendrogram(clusters,labels<span class="op">=</span>labels,leaf_rotation<span class="op">=</span><span class="dv">90</span>)</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-clustering-example-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-04-03-binding-site-comparison-benchmark-I_files/figure-html/fig-clustering-example-1-output-1.png" width="571" height="438" class="figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;10: Dendrogram from single linkage clustering of previous distance matrix</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>Now that I have this, I can cut the result at any given distance and get the clusters. Obviously this example is a bit too tiny to draw any interesting insights. So in order to get ready for our first full dataset, we need to repeat the whole process for all the structures weâ€™ve identified earlier. In order to do so I first need to properly address two items on my todo list that I tracked as side notes in the previous sections (Iâ€™ll put it here again as reference).</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Things to keep in mind</p>
<ul>
<li>Uniprot accession vs polymer entity group ID handling to be clarified with RCSB folks</li>
<li>Donâ€™t forget to check if all sidechains of the binding site are resolved in the selected structures</li>
<li>Treat homo-multimers rationally in the data-set construction</li>
<li>Handle canonical residue numbering properly in distanceMatrix code</li>
<li>Consider also optional side-chain sensitive distance calculations</li>
<li>What to do with large motions implicating new residues in binding sites?</li>
</ul>
</div></div><ol type="1">
<li>I need to verify that all structures have a canonical uniprot sequence residue numbering. If not I need to map the residue numbers to the canonical uniprot sequence.</li>
<li>I need to know which chain I use beforehand (7l7j contains two different proteins for instance - so just sticking without a selection is not an option)</li>
</ol>
</section>
</section>
<section id="automating-things" class="level3 page-columns page-full">
<h3 class="anchored" data-anchor-id="automating-things">Automating things</h3>
<p>The first issue previously mentioned was briefly checked against 3decision, where I could see that none of the RCSB structures have sequence mapping. So in the case of the selected HSP90 structures and this binding site, itâ€™s not an issue to perform residue selections using the residue numbers I used here. You could extract that kind of info using a graphql query from the RCSB as well, but Iâ€™ll pass here and will come back to that if / when needed. In order to address the second of the previously mentioned issue I am going to adjust a bit the graphql query I used earlier and track for each structure which chain corresponds to HSP90 containing the binding site we are interested in here.</p>
<div class="cell" data-execution_count="12">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># here we can reuse the already retrieved data from the GraphQL query that we did earlier against the RCSB, because it already contains the information we need.</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="co"># a bit of rewrite of the function we had before (still very ugly, feel free to make it pretty ;) </span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> extractSequenceRegionsAndChains(data):</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  result<span class="op">=</span>{}</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> entry <span class="kw">in</span> data:  </span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>    result[entry[<span class="st">"rcsb_id"</span>]]<span class="op">=</span>[]</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> entity <span class="kw">in</span> entry[<span class="st">"polymer_entities"</span>]:</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>      <span class="co"># print(entity["rcsb_polymer_entity_container_identifiers"])</span></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> entity[<span class="st">"entity_poly"</span>][<span class="st">"rcsb_mutation_count"</span>]<span class="op">==</span><span class="dv">0</span>:</span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>        chain_ids<span class="op">=</span>entity[<span class="st">"rcsb_polymer_entity_container_identifiers"</span>][<span class="st">"auth_asym_ids"</span>]</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> idx,identifier <span class="kw">in</span> <span class="bu">enumerate</span>(entity[<span class="st">"rcsb_polymer_entity_container_identifiers"</span>][<span class="st">"reference_sequence_identifiers"</span>]):</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>          <span class="cf">if</span> identifier[<span class="st">"database_name"</span>]<span class="op">==</span><span class="st">"UniProt"</span> <span class="kw">and</span> identifier[<span class="st">"database_accession"</span>]<span class="op">==</span><span class="st">"P07900"</span>:</span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> entity[<span class="st">"rcsb_polymer_entity_align"</span>][idx][<span class="st">"reference_database_name"</span>]<span class="op">==</span><span class="st">"UniProt"</span>:</span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>              <span class="cf">for</span> region <span class="kw">in</span> entity[<span class="st">"rcsb_polymer_entity_align"</span>][idx][<span class="st">"aligned_regions"</span>]:</span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>                result[entry[<span class="st">"rcsb_id"</span>]].append((region[<span class="st">"ref_beg_seq_id"</span>],region[<span class="st">"ref_beg_seq_id"</span>]<span class="op">+</span>region[<span class="st">"length"</span>]))</span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a>                result[entry[<span class="st">"rcsb_id"</span>]].append(chain_ids)</span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> result</span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a><span class="co"># residuelist=[47,48,51,52,54,55,58,93,95,96,97,98,106,107,112,131,132,133,134,135,136,137,138,139,152,183,184,186]</span></span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a>resolved_regions<span class="op">=</span>extractSequenceRegionsAndChains(dataRCSBGQL)</span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a>selectedPdbCodesAndChains<span class="op">=</span>[]</span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> pdbCode <span class="kw">in</span> resolved_regions:</span>
<span id="cb45-29"><a href="#cb45-29" aria-hidden="true" tabindex="-1"></a>  mapping <span class="op">=</span> resolved_regions[pdbCode]</span>
<span id="cb45-30"><a href="#cb45-30" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="bu">len</span>(mapping)<span class="op">&gt;</span><span class="dv">0</span> <span class="kw">and</span> <span class="bu">all</span>(i<span class="op">&gt;=</span>mapping[<span class="dv">0</span>][<span class="dv">0</span>] <span class="kw">and</span> i<span class="op">&lt;=</span>mapping[<span class="dv">0</span>][<span class="dv">1</span>] <span class="cf">for</span> i <span class="kw">in</span> residuelist)):</span>
<span id="cb45-31"><a href="#cb45-31" aria-hidden="true" tabindex="-1"></a>    selectedPdbCodesAndChains.append({<span class="st">"pdbCode"</span>:pdbCode,<span class="st">"chain"</span>:mapping[<span class="dv">1</span>][<span class="dv">0</span>]})</span>
<span id="cb45-32"><a href="#cb45-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-33"><a href="#cb45-33" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(selectedPdbCodesAndChains)</span>
<span id="cb45-34"><a href="#cb45-34" aria-hidden="true" tabindex="-1"></a><span class="co"># print(str(len(selectedPdbCodes)) + " structures identified")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[{'pdbCode': '1BYQ', 'chain': 'A'}, {'pdbCode': '1OSF', 'chain': 'A'}, {'pdbCode': '1UY6', 'chain': 'A'}, {'pdbCode': '1UY7', 'chain': 'A'}, {'pdbCode': '1UY8', 'chain': 'A'}, {'pdbCode': '1UY9', 'chain': 'A'}, {'pdbCode': '1UYC', 'chain': 'A'}, {'pdbCode': '1UYD', 'chain': 'A'}, {'pdbCode': '1UYE', 'chain': 'A'}, {'pdbCode': '1UYF', 'chain': 'A'}, {'pdbCode': '1UYG', 'chain': 'A'}, {'pdbCode': '1UYH', 'chain': 'A'}, {'pdbCode': '1UYI', 'chain': 'A'}, {'pdbCode': '1UYK', 'chain': 'A'}, {'pdbCode': '1UYL', 'chain': 'A'}, {'pdbCode': '1YC1', 'chain': 'A'}, {'pdbCode': '1YC3', 'chain': 'A'}, {'pdbCode': '1YC4', 'chain': 'A'}, {'pdbCode': '1YER', 'chain': 'A'}, {'pdbCode': '1YES', 'chain': 'A'}, {'pdbCode': '1YET', 'chain': 'A'}, {'pdbCode': '2BSM', 'chain': 'A'}, {'pdbCode': '2BT0', 'chain': 'A'}, {'pdbCode': '2BYH', 'chain': 'A'}, {'pdbCode': '2BYI', 'chain': 'A'}, {'pdbCode': '2BZ5', 'chain': 'A'}, {'pdbCode': '2CCS', 'chain': 'A'}, {'pdbCode': '2CCT', 'chain': 'A'}, {'pdbCode': '2CCU', 'chain': 'A'}, {'pdbCode': '2FWY', 'chain': 'A'}, {'pdbCode': '2FWZ', 'chain': 'A'}, {'pdbCode': '2H55', 'chain': 'A'}, {'pdbCode': '2JJC', 'chain': 'A'}, {'pdbCode': '2K5B', 'chain': 'A'}, {'pdbCode': '2QF6', 'chain': 'A'}, {'pdbCode': '2QFO', 'chain': 'A'}, {'pdbCode': '2QG0', 'chain': 'A'}, {'pdbCode': '2QG2', 'chain': 'A'}, {'pdbCode': '2UWD', 'chain': 'A'}, {'pdbCode': '2VCI', 'chain': 'A'}, {'pdbCode': '2VCJ', 'chain': 'A'}, {'pdbCode': '2WI1', 'chain': 'A'}, {'pdbCode': '2WI2', 'chain': 'A'}, {'pdbCode': '2WI3', 'chain': 'A'}, {'pdbCode': '2WI4', 'chain': 'A'}, {'pdbCode': '2WI5', 'chain': 'A'}, {'pdbCode': '2WI6', 'chain': 'A'}, {'pdbCode': '2WI7', 'chain': 'A'}, {'pdbCode': '2XAB', 'chain': 'A'}, {'pdbCode': '2XDK', 'chain': 'A'}, {'pdbCode': '2XDL', 'chain': 'A'}, {'pdbCode': '2XDS', 'chain': 'A'}, {'pdbCode': '2XDU', 'chain': 'A'}, {'pdbCode': '2XDX', 'chain': 'A'}, {'pdbCode': '2XHR', 'chain': 'A'}, {'pdbCode': '2XHT', 'chain': 'A'}, {'pdbCode': '2XHX', 'chain': 'A'}, {'pdbCode': '2XJG', 'chain': 'A'}, {'pdbCode': '2XJJ', 'chain': 'A'}, {'pdbCode': '2XJX', 'chain': 'A'}, {'pdbCode': '2XK2', 'chain': 'A'}, {'pdbCode': '2YE2', 'chain': 'A'}, {'pdbCode': '2YE3', 'chain': 'A'}, {'pdbCode': '2YE4', 'chain': 'A'}, {'pdbCode': '2YE5', 'chain': 'A'}, {'pdbCode': '2YE6', 'chain': 'A'}, {'pdbCode': '2YE7', 'chain': 'A'}, {'pdbCode': '2YE8', 'chain': 'A'}, {'pdbCode': '2YE9', 'chain': 'A'}, {'pdbCode': '2YEA', 'chain': 'A'}, {'pdbCode': '2YEB', 'chain': 'A'}, {'pdbCode': '2YEC', 'chain': 'A'}, {'pdbCode': '2YED', 'chain': 'A'}, {'pdbCode': '2YEE', 'chain': 'A'}, {'pdbCode': '2YEF', 'chain': 'A'}, {'pdbCode': '2YEG', 'chain': 'A'}, {'pdbCode': '2YEH', 'chain': 'A'}, {'pdbCode': '2YEI', 'chain': 'A'}, {'pdbCode': '2YEJ', 'chain': 'A'}, {'pdbCode': '2YI0', 'chain': 'A'}, {'pdbCode': '2YI5', 'chain': 'A'}, {'pdbCode': '2YI6', 'chain': 'A'}, {'pdbCode': '2YI7', 'chain': 'A'}, {'pdbCode': '2YJW', 'chain': 'A'}, {'pdbCode': '2YJX', 'chain': 'A'}, {'pdbCode': '2YK2', 'chain': 'A'}, {'pdbCode': '2YK9', 'chain': 'A'}, {'pdbCode': '2YKB', 'chain': 'A'}, {'pdbCode': '2YKC', 'chain': 'A'}, {'pdbCode': '2YKE', 'chain': 'A'}, {'pdbCode': '2YKI', 'chain': 'A'}, {'pdbCode': '2YKJ', 'chain': 'A'}, {'pdbCode': '3B24', 'chain': 'A'}, {'pdbCode': '3B25', 'chain': 'A'}, {'pdbCode': '3B26', 'chain': 'A'}, {'pdbCode': '3B27', 'chain': 'A'}, {'pdbCode': '3B28', 'chain': 'A'}, {'pdbCode': '3BM9', 'chain': 'A'}, {'pdbCode': '3BMY', 'chain': 'A'}, {'pdbCode': '3D0B', 'chain': 'A'}, {'pdbCode': '3EKO', 'chain': 'A'}, {'pdbCode': '3EKR', 'chain': 'A'}, {'pdbCode': '3FT5', 'chain': 'A'}, {'pdbCode': '3FT8', 'chain': 'A'}, {'pdbCode': '3HEK', 'chain': 'A'}, {'pdbCode': '3HHU', 'chain': 'A'}, {'pdbCode': '3HYY', 'chain': 'A'}, {'pdbCode': '3HYZ', 'chain': 'A'}, {'pdbCode': '3HZ1', 'chain': 'A'}, {'pdbCode': '3HZ5', 'chain': 'A'}, {'pdbCode': '3INW', 'chain': 'A'}, {'pdbCode': '3INX', 'chain': 'A'}, {'pdbCode': '3K97', 'chain': 'A'}, {'pdbCode': '3K98', 'chain': 'A'}, {'pdbCode': '3K99', 'chain': 'A'}, {'pdbCode': '3MNR', 'chain': 'P'}, {'pdbCode': '3O0I', 'chain': 'A'}, {'pdbCode': '3OW6', 'chain': 'A'}, {'pdbCode': '3OWB', 'chain': 'A'}, {'pdbCode': '3OWD', 'chain': 'A'}, {'pdbCode': '3QDD', 'chain': 'A'}, {'pdbCode': '3QTF', 'chain': 'A'}, {'pdbCode': '3R4M', 'chain': 'A'}, {'pdbCode': '3R4N', 'chain': 'A'}, {'pdbCode': '3R4O', 'chain': 'A'}, {'pdbCode': '3R4P', 'chain': 'A'}, {'pdbCode': '3R91', 'chain': 'A'}, {'pdbCode': '3R92', 'chain': 'A'}, {'pdbCode': '3RKZ', 'chain': 'A'}, {'pdbCode': '3RLP', 'chain': 'A'}, {'pdbCode': '3RLQ', 'chain': 'A'}, {'pdbCode': '3RLR', 'chain': 'A'}, {'pdbCode': '3T0H', 'chain': 'A'}, {'pdbCode': '3T0Z', 'chain': 'A'}, {'pdbCode': '3T10', 'chain': 'A'}, {'pdbCode': '3T1K', 'chain': 'A'}, {'pdbCode': '3T2S', 'chain': 'A'}, {'pdbCode': '3TUH', 'chain': 'A'}, {'pdbCode': '3VHA', 'chain': 'A'}, {'pdbCode': '3VHC', 'chain': 'A'}, {'pdbCode': '3VHD', 'chain': 'A'}, {'pdbCode': '3WHA', 'chain': 'A'}, {'pdbCode': '3WQ9', 'chain': 'A'}, {'pdbCode': '4AWO', 'chain': 'A'}, {'pdbCode': '4AWP', 'chain': 'A'}, {'pdbCode': '4AWQ', 'chain': 'A'}, {'pdbCode': '4B7P', 'chain': 'A'}, {'pdbCode': '4BQG', 'chain': 'A'}, {'pdbCode': '4BQJ', 'chain': 'A'}, {'pdbCode': '4CWF', 'chain': 'A'}, {'pdbCode': '4CWN', 'chain': 'A'}, {'pdbCode': '4CWO', 'chain': 'A'}, {'pdbCode': '4CWP', 'chain': 'A'}, {'pdbCode': '4CWQ', 'chain': 'A'}, {'pdbCode': '4CWR', 'chain': 'A'}, {'pdbCode': '4CWS', 'chain': 'A'}, {'pdbCode': '4CWT', 'chain': 'A'}, {'pdbCode': '4EEH', 'chain': 'A'}, {'pdbCode': '4EFT', 'chain': 'A'}, {'pdbCode': '4EFU', 'chain': 'A'}, {'pdbCode': '4EGH', 'chain': 'A'}, {'pdbCode': '4EGI', 'chain': 'A'}, {'pdbCode': '4EGK', 'chain': 'A'}, {'pdbCode': '4FCP', 'chain': 'A'}, {'pdbCode': '4FCQ', 'chain': 'A'}, {'pdbCode': '4FCR', 'chain': 'A'}, {'pdbCode': '4HY6', 'chain': 'A'}, {'pdbCode': '4JQL', 'chain': 'A'}, {'pdbCode': '4L8Z', 'chain': 'A'}, {'pdbCode': '4L90', 'chain': 'A'}, {'pdbCode': '4L91', 'chain': 'A'}, {'pdbCode': '4L93', 'chain': 'A'}, {'pdbCode': '4L94', 'chain': 'A'}, {'pdbCode': '4LWE', 'chain': 'A'}, {'pdbCode': '4LWF', 'chain': 'A'}, {'pdbCode': '4LWG', 'chain': 'A'}, {'pdbCode': '4LWH', 'chain': 'A'}, {'pdbCode': '4LWI', 'chain': 'A'}, {'pdbCode': '4NH7', 'chain': 'A'}, {'pdbCode': '4NH8', 'chain': 'A'}, {'pdbCode': '4O04', 'chain': 'A'}, {'pdbCode': '4O05', 'chain': 'A'}, {'pdbCode': '4O07', 'chain': 'A'}, {'pdbCode': '4O09', 'chain': 'A'}, {'pdbCode': '4O0B', 'chain': 'A'}, {'pdbCode': '4R3M', 'chain': 'A'}, {'pdbCode': '4U93', 'chain': 'A'}, {'pdbCode': '4W7T', 'chain': 'A'}, {'pdbCode': '4XIP', 'chain': 'A'}, {'pdbCode': '4XIQ', 'chain': 'A'}, {'pdbCode': '4XIR', 'chain': 'A'}, {'pdbCode': '4XIT', 'chain': 'A'}, {'pdbCode': '4YKQ', 'chain': 'A'}, {'pdbCode': '4YKR', 'chain': 'A'}, {'pdbCode': '4YKT', 'chain': 'A'}, {'pdbCode': '4YKU', 'chain': 'A'}, {'pdbCode': '4YKW', 'chain': 'A'}, {'pdbCode': '4YKX', 'chain': 'A'}, {'pdbCode': '4YKY', 'chain': 'A'}, {'pdbCode': '4YKZ', 'chain': 'A'}, {'pdbCode': '5CF0', 'chain': 'A'}, {'pdbCode': '5FNC', 'chain': 'A'}, {'pdbCode': '5FND', 'chain': 'A'}, {'pdbCode': '5FNF', 'chain': 'A'}, {'pdbCode': '5GGZ', 'chain': 'A'}, {'pdbCode': '5J20', 'chain': 'A'}, {'pdbCode': '5J27', 'chain': 'A'}, {'pdbCode': '5J2V', 'chain': 'A'}, {'pdbCode': '5J2X', 'chain': 'A'}, {'pdbCode': '5J64', 'chain': 'A'}, {'pdbCode': '5J6L', 'chain': 'A'}, {'pdbCode': '5J6M', 'chain': 'A'}, {'pdbCode': '5J6N', 'chain': 'A'}, {'pdbCode': '5J82', 'chain': 'A'}, {'pdbCode': '5J86', 'chain': 'A'}, {'pdbCode': '5J8M', 'chain': 'A'}, {'pdbCode': '5J8U', 'chain': 'A'}, {'pdbCode': '5J9X', 'chain': 'A'}, {'pdbCode': '5LNY', 'chain': 'A'}, {'pdbCode': '5LNZ', 'chain': 'A'}, {'pdbCode': '5LO0', 'chain': 'A'}, {'pdbCode': '5LO1', 'chain': 'A'}, {'pdbCode': '5LO5', 'chain': 'A'}, {'pdbCode': '5LO6', 'chain': 'A'}, {'pdbCode': '5LQ9', 'chain': 'A'}, {'pdbCode': '5LR1', 'chain': 'A'}, {'pdbCode': '5LR7', 'chain': 'A'}, {'pdbCode': '5LRL', 'chain': 'A'}, {'pdbCode': '5LRZ', 'chain': 'A'}, {'pdbCode': '5LS1', 'chain': 'A'}, {'pdbCode': '5M4E', 'chain': 'A'}, {'pdbCode': '5M4H', 'chain': 'A'}, {'pdbCode': '5NYH', 'chain': 'A'}, {'pdbCode': '5NYI', 'chain': 'A'}, {'pdbCode': '5OCI', 'chain': 'A'}, {'pdbCode': '5OD7', 'chain': 'A'}, {'pdbCode': '5ODX', 'chain': 'A'}, {'pdbCode': '5T21', 'chain': 'A'}, {'pdbCode': '5VYY', 'chain': 'A'}, {'pdbCode': '5XQD', 'chain': 'A'}, {'pdbCode': '5XQE', 'chain': 'A'}, {'pdbCode': '5XR5', 'chain': 'A'}, {'pdbCode': '5XR9', 'chain': 'A'}, {'pdbCode': '5XRB', 'chain': 'A'}, {'pdbCode': '5XRD', 'chain': 'A'}, {'pdbCode': '5XRE', 'chain': 'A'}, {'pdbCode': '5ZR3', 'chain': 'A'}, {'pdbCode': '6B99', 'chain': 'A'}, {'pdbCode': '6B9A', 'chain': 'A'}, {'pdbCode': '6CEO', 'chain': 'A'}, {'pdbCode': '6CYG', 'chain': 'A'}, {'pdbCode': '6CYH', 'chain': 'A'}, {'pdbCode': '6EI5', 'chain': 'A'}, {'pdbCode': '6EL5', 'chain': 'A'}, {'pdbCode': '6ELN', 'chain': 'A'}, {'pdbCode': '6ELO', 'chain': 'A'}, {'pdbCode': '6ELP', 'chain': 'A'}, {'pdbCode': '6EY8', 'chain': 'A'}, {'pdbCode': '6EY9', 'chain': 'A'}, {'pdbCode': '6EYA', 'chain': 'A'}, {'pdbCode': '6EYB', 'chain': 'A'}, {'pdbCode': '6F1N', 'chain': 'A'}, {'pdbCode': '6FCJ', 'chain': 'A'}, {'pdbCode': '6GPR', 'chain': 'A'}, {'pdbCode': '6GPT', 'chain': 'A'}, {'pdbCode': '6GPW', 'chain': 'A'}, {'pdbCode': '6GPY', 'chain': 'A'}, {'pdbCode': '6GQ6', 'chain': 'A'}, {'pdbCode': '6GQR', 'chain': 'A'}, {'pdbCode': '6HHR', 'chain': 'A'}, {'pdbCode': '6LR9', 'chain': 'A'}, {'pdbCode': '6LSZ', 'chain': 'A'}, {'pdbCode': '6LT8', 'chain': 'A'}, {'pdbCode': '6LTI', 'chain': 'A'}, {'pdbCode': '6LTK', 'chain': 'A'}, {'pdbCode': '6N8X', 'chain': 'A'}, {'pdbCode': '6TN4', 'chain': 'AAA'}, {'pdbCode': '6TN5', 'chain': 'AAA'}, {'pdbCode': '6U99', 'chain': 'A'}, {'pdbCode': '6U9B', 'chain': 'A'}, {'pdbCode': '7DMC', 'chain': 'A'}, {'pdbCode': '7KRJ', 'chain': 'A'}, {'pdbCode': '7KW7', 'chain': 'A'}, {'pdbCode': '7L7I', 'chain': 'A'}, {'pdbCode': '7L7J', 'chain': 'A'}, {'pdbCode': '7LSZ', 'chain': 'A'}, {'pdbCode': '7LT0', 'chain': 'A'}, {'pdbCode': '7S8Y', 'chain': 'A'}, {'pdbCode': '7S8Z', 'chain': 'A'}, {'pdbCode': '7S90', 'chain': 'A'}, {'pdbCode': '7S95', 'chain': 'A'}, {'pdbCode': '7S98', 'chain': 'A'}, {'pdbCode': '7S99', 'chain': 'A'}, {'pdbCode': '7S9F', 'chain': 'A'}, {'pdbCode': '7S9G', 'chain': 'A'}, {'pdbCode': '7S9H', 'chain': 'A'}, {'pdbCode': '7S9I', 'chain': 'A'}, {'pdbCode': '7UR3', 'chain': 'A'}, {'pdbCode': '8AGI', 'chain': 'A'}]</code></pre>
</div>
</div>
<p>Now we have the list of pdb codes &amp; the chain we need to use here. Letâ€™s wrap up the rest of the whole process to produce a larger distance matrix &amp; clustering result. In the list of pdb codes above several structures are NOT anymore available in the PDB format. So I have to adapt the previous code to work with mmCIF files instead. And lucky as I am, MDanalysis doesnâ€™t seem to support them very well, nor mmtf which apparently doesnâ€™t read in chain information from mmCIF files. So Iâ€™ll have to rewrite all of that using a different toolkit closer to crystallographers, i.e.&nbsp;<a href="https://github.com/project-gemmi/gemmi">gemmi</a>.</p>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> gemmi <span class="im">import</span> cif</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> urllib</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>pdbCode<span class="op">=</span><span class="st">"3t0z"</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>chainCode<span class="op">=</span><span class="st">"A"</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>content<span class="op">=</span> urllib.request.urlopen(<span class="st">"https://files.rcsb.org/view/"</span><span class="op">+</span>pdbCode<span class="op">+</span><span class="st">".cif"</span>).read()</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>block<span class="op">=</span>cif.read_string(content)[<span class="dv">0</span>]</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>structure<span class="op">=</span>gemmi.make_structure_from_block(block)</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>positions<span class="op">=</span>[]</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> model <span class="kw">in</span> structure:</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> chain <span class="kw">in</span> model:</span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> chain.name <span class="op">==</span> chainCode:</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> residue <span class="kw">in</span> chain:</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> residue.seqid.num <span class="kw">in</span> residuelist:</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>          <span class="cf">for</span> atom <span class="kw">in</span> residue:</span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> atom.name<span class="op">==</span><span class="st">"CA"</span>:</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>              positions.append(atom.pos.tolist())</span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="bu">len</span>(positions)<span class="op">!=</span><span class="bu">len</span>(residuelist)):</span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"Not all positions found"</span>)</span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"stopping here"</span>)</span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a>positions_np<span class="op">=</span>np.array(positions)</span>
<span id="cb47-26"><a href="#cb47-26" aria-hidden="true" tabindex="-1"></a>distances3t0zgemmi <span class="op">=</span> cdist(positions_np, positions_np, <span class="st">'euclidean'</span>)</span>
<span id="cb47-27"><a href="#cb47-27" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(np.<span class="bu">all</span>(distances3t0z<span class="op">-</span>distances3t0zgemmi<span class="op">&lt;</span><span class="fl">0.0001</span>)):</span>
<span id="cb47-28"><a href="#cb47-28" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"distances match with previous analysis"</span>)</span>
<span id="cb47-29"><a href="#cb47-29" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb47-30"><a href="#cb47-30" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"distances don't match with previous analysis"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>distances match with previous analysis</code></pre>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>Things to keep in mind</p>
<ul>
<li>Uniprot accession vs polymer entity group ID handling to be clarified with RCSB folks</li>
<li>Donâ€™t forget to check if all sidechains of the binding site are resolved in the selected structures</li>
<li>Treat homo-multimers rationally in the data-set construction</li>
<li>Handle canonical residue numbering properly in distanceMatrix code</li>
<li>Consider also optional side-chain sensitive distance calculations</li>
<li>What to do with large motions implicating new residues in binding sites?</li>
<li>Decide how to handle NMR conformations in the dataset</li>
</ul>
</div></div><p>In the code snippet above Iâ€™m checking if the results we obtained with the horrible quadruple loop match the results we got earlier with MDAnalysis on the same structure. This appears to be the case here.</p>
<p>Now letâ€™s put the whole thing into a function, loop over all our structures &amp; I guess adjust what needs adjustments (as for sure Iâ€™ll have something that will fail <span class="emoji" data-emoji="tired_face">ðŸ˜«</span>). By the time of the writing 4 structures were found to have 1 residue implicated in the binding site to be not resolved. Theyâ€™ll be discarded from further analyses.</p>
<div class="cell" data-execution_count="14">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> getContactMatrix(pdbCode, chainCode, residueSelection, debug<span class="op">=</span><span class="va">False</span>):</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  content<span class="op">=</span> urllib.request.urlopen(<span class="st">"https://files.rcsb.org/view/"</span><span class="op">+</span>pdbCode<span class="op">+</span><span class="st">".cif"</span>).read()</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  block<span class="op">=</span>cif.read_string(content)[<span class="dv">0</span>]</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  structure<span class="op">=</span>gemmi.make_structure_from_block(block)</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>  positions<span class="op">=</span>[]</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> model <span class="kw">in</span> structure:</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> model.name<span class="op">==</span><span class="st">"1"</span>:</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> chain <span class="kw">in</span> model:</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> chain.name <span class="op">==</span> chainCode:</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>          <span class="cf">for</span> residue <span class="kw">in</span> chain:</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> residue.seqid.num <span class="kw">in</span> residueSelection:</span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a>              <span class="cf">if</span> debug: <span class="bu">print</span>(residue.seqid.num)</span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a>              <span class="cf">for</span> atom <span class="kw">in</span> residue:</span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> atom.name<span class="op">==</span><span class="st">"CA"</span>:</span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a>                  <span class="cf">if</span> debug: <span class="bu">print</span>(<span class="st">"ok"</span>)</span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a>                  positions.append(atom.pos.tolist())</span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a>                  <span class="cf">break</span> <span class="co">#we need that for multiple occurences</span></span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="bu">len</span>(positions)<span class="op">!=</span><span class="bu">len</span>(residueSelection)):</span>
<span id="cb49-22"><a href="#cb49-22" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Not all positions found for "</span><span class="op">+</span>pdbCode<span class="op">+</span><span class="st">" discarding structure"</span>)</span>
<span id="cb49-23"><a href="#cb49-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="va">None</span></span>
<span id="cb49-24"><a href="#cb49-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-25"><a href="#cb49-25" aria-hidden="true" tabindex="-1"></a>  positions_np<span class="op">=</span>np.array(positions)</span>
<span id="cb49-26"><a href="#cb49-26" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> cdist(positions_np, positions_np, <span class="st">'euclidean'</span>)</span>
<span id="cb49-27"><a href="#cb49-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-28"><a href="#cb49-28" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> clusterMatrices(matrixList):  </span>
<span id="cb49-29"><a href="#cb49-29" aria-hidden="true" tabindex="-1"></a>  nMatrices<span class="op">=</span><span class="bu">len</span>(matrixList)</span>
<span id="cb49-30"><a href="#cb49-30" aria-hidden="true" tabindex="-1"></a>  result<span class="op">=</span>np.zeros((nMatrices,nMatrices))</span>
<span id="cb49-31"><a href="#cb49-31" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> i <span class="kw">in</span> <span class="bu">range</span>(nMatrices):</span>
<span id="cb49-32"><a href="#cb49-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> j <span class="kw">in</span> <span class="bu">range</span>(nMatrices):</span>
<span id="cb49-33"><a href="#cb49-33" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> i<span class="op">==</span>j:</span>
<span id="cb49-34"><a href="#cb49-34" aria-hidden="true" tabindex="-1"></a>        result[i][j]<span class="op">=</span><span class="fl">0.0</span></span>
<span id="cb49-35"><a href="#cb49-35" aria-hidden="true" tabindex="-1"></a>      <span class="cf">elif</span> i<span class="op">&lt;</span>j:</span>
<span id="cb49-36"><a href="#cb49-36" aria-hidden="true" tabindex="-1"></a>        result[i][j]<span class="op">=</span>np.mean(np.<span class="bu">abs</span>(matrixList[i]<span class="op">-</span>matrixList[j]))</span>
<span id="cb49-37"><a href="#cb49-37" aria-hidden="true" tabindex="-1"></a>        result[j][i]<span class="op">=</span>result[i][j]</span>
<span id="cb49-38"><a href="#cb49-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-39"><a href="#cb49-39" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb49-40"><a href="#cb49-40" aria-hidden="true" tabindex="-1"></a>  distArray <span class="op">=</span> ssd.squareform(result)</span>
<span id="cb49-41"><a href="#cb49-41" aria-hidden="true" tabindex="-1"></a>  clusters<span class="op">=</span>scipy.cluster.hierarchy.linkage(result, method<span class="op">=</span><span class="st">'single'</span>, metric<span class="op">=</span><span class="st">'euclidean'</span>)</span>
<span id="cb49-42"><a href="#cb49-42" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span>(clusters)</span>
<span id="cb49-43"><a href="#cb49-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-44"><a href="#cb49-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-45"><a href="#cb49-45" aria-hidden="true" tabindex="-1"></a>contactMatrices<span class="op">=</span>[getContactMatrix(structure[<span class="st">"pdbCode"</span>],structure[<span class="st">"chain"</span>],residuelist) <span class="cf">for</span> structure <span class="kw">in</span> selectedPdbCodesAndChains]</span>
<span id="cb49-46"><a href="#cb49-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-47"><a href="#cb49-47" aria-hidden="true" tabindex="-1"></a>none_indices <span class="op">=</span> [ic <span class="cf">for</span> ic, matrix <span class="kw">in</span> <span class="bu">enumerate</span>(contactMatrices) <span class="cf">if</span> matrix <span class="kw">is</span> <span class="va">None</span>]</span>
<span id="cb49-48"><a href="#cb49-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-49"><a href="#cb49-49" aria-hidden="true" tabindex="-1"></a>labels<span class="op">=</span>[structure[<span class="st">"pdbCode"</span>] <span class="cf">for</span> structure <span class="kw">in</span> selectedPdbCodesAndChains]</span>
<span id="cb49-50"><a href="#cb49-50" aria-hidden="true" tabindex="-1"></a>chainCodes<span class="op">=</span>[structure[<span class="st">"chain"</span>] <span class="cf">for</span> structure <span class="kw">in</span> selectedPdbCodesAndChains]</span>
<span id="cb49-51"><a href="#cb49-51" aria-hidden="true" tabindex="-1"></a>filteredContactMatrices <span class="op">=</span> [matrix <span class="cf">for</span> i, matrix <span class="kw">in</span> <span class="bu">enumerate</span>(contactMatrices) <span class="cf">if</span> i <span class="kw">not</span> <span class="kw">in</span> none_indices]</span>
<span id="cb49-52"><a href="#cb49-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-53"><a href="#cb49-53" aria-hidden="true" tabindex="-1"></a>filteredLabels <span class="op">=</span> [label <span class="cf">for</span> idx, label <span class="kw">in</span> <span class="bu">enumerate</span>(labels) <span class="cf">if</span> idx <span class="kw">not</span> <span class="kw">in</span> none_indices]</span>
<span id="cb49-54"><a href="#cb49-54" aria-hidden="true" tabindex="-1"></a>filteredChainCodes<span class="op">=</span> [code <span class="cf">for</span> idx, code <span class="kw">in</span> <span class="bu">enumerate</span>(chainCodes) <span class="cf">if</span> idx <span class="kw">not</span> <span class="kw">in</span> none_indices]</span>
<span id="cb49-55"><a href="#cb49-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-56"><a href="#cb49-56" aria-hidden="true" tabindex="-1"></a>clusters<span class="op">=</span>clusterMatrices(filteredContactMatrices)</span>
<span id="cb49-57"><a href="#cb49-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-58"><a href="#cb49-58" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">40</span>))</span>
<span id="cb49-59"><a href="#cb49-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-60"><a href="#cb49-60" aria-hidden="true" tabindex="-1"></a>scipy.cluster.hierarchy.dendrogram(clusters,labels<span class="op">=</span>filteredLabels,orientation<span class="op">=</span><span class="st">'right'</span>,leaf_font_size<span class="op">=</span><span class="dv">7</span>,color_threshold<span class="op">=</span><span class="fl">1.0</span>)</span>
<span id="cb49-61"><a href="#cb49-61" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Not all positions found for 2YI5 discarding structure
Not all positions found for 3HYZ discarding structure
Not all positions found for 4FCR discarding structure
Not all positions found for 6EYA discarding structure</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/var/folders/51/p06ylvs507q40bm1kxbl65km0000gp/T/ipykernel_19251/1742338535.py:41: ClusterWarning:

scipy.cluster: The symmetric non-negative hollow observation matrix looks suspiciously like an uncondensed distance matrix
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-clustering-final" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-04-03-binding-site-comparison-benchmark-I_files/figure-html/fig-clustering-final-output-3.png" width="802" height="2998" class="figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;11: Result of hierarchical clustering of HSP90 ATP binding sites based on CA positions</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>Alright, this is starting to look like something useable. The colors encode clusters, when cutting the dendrogram at a distance of 1.0 we get 3 very large clusters and several smaller ones &amp; outliers.</p>
<p>Letâ€™s cherrypick 2 structures from each of the 3 large clusters and see if we can observe structural reasons underlinging these results:</p>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb52" data-startfrom="967" data-source-offset="0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 966;"><span id="cb52-967"><a href="#cb52-967" aria-hidden="true" tabindex="-1"></a><span class="co">// Create drawing area</span></span>
<span id="cb52-968"><a href="#cb52-968" aria-hidden="true" tabindex="-1"></a>divNGL5 <span class="op">=</span> <span class="fu">html</span><span class="vs">`&lt;div style="width:800px;height:600px;position:relative"&gt;&lt;/div&gt;`</span><span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-hsp-5" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div id="ojs-cell-9" data-nodetype="declaration">

</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;12: Comparison of 2 representative binding sites for each of the 3 large clusters identified. Orange, 2 structures from the orange cluster of <a href="#fig-clustering-final">Figure&nbsp;11</a>; red, 2 structures from the red cluster of <a href="#fig-clustering-final">Figure&nbsp;11</a>; blue, 2 structures from the brown cluster of <a href="#fig-clustering-final">Figure&nbsp;11</a>.</figcaption><p></p>
</figure>
</div>
</div>
<div class="cell">
<details class="hidden">
<summary>Code</summary>
<div class="sourceCode cell-code hidden" id="cb53" data-startfrom="977" data-source-offset="-0"><pre class="sourceCode js code-with-copy"><code class="sourceCode javascript" style="counter-reset: source-line 976;"><span id="cb53-977"><a href="#cb53-977" aria-hidden="true" tabindex="-1"></a>stage5 <span class="op">=</span> <span class="kw">new</span> NGL<span class="op">.</span><span class="fu">Stage</span>(divNGL5<span class="op">,</span> { <span class="dt">backgroundColor</span><span class="op">:</span> <span class="st">"black"</span><span class="op">,</span><span class="dt">cameraType</span><span class="op">:</span> <span class="st">"orthographic"</span>})<span class="op">;</span></span>
<span id="cb53-978"><a href="#cb53-978" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb53-979"><a href="#cb53-979" aria-hidden="true" tabindex="-1"></a>  <span class="kw">let</span> files<span class="op">=</span>[<span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/4yku_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">,</span></span>
<span id="cb53-980"><a href="#cb53-980" aria-hidden="true" tabindex="-1"></a>    <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/2yi6_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">,</span></span>
<span id="cb53-981"><a href="#cb53-981" aria-hidden="true" tabindex="-1"></a>    <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/1yc4_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">,</span></span>
<span id="cb53-982"><a href="#cb53-982" aria-hidden="true" tabindex="-1"></a>    <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/2xjg_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">,</span></span>
<span id="cb53-983"><a href="#cb53-983" aria-hidden="true" tabindex="-1"></a>    <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/4o05_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()<span class="op">,</span></span>
<span id="cb53-984"><a href="#cb53-984" aria-hidden="true" tabindex="-1"></a>    <span class="cf">await</span> <span class="fu">FileAttachment</span>(<span class="st">"pdbs/2xdu_aligned.pdb"</span>)<span class="op">.</span><span class="fu">blob</span>()]</span>
<span id="cb53-985"><a href="#cb53-985" aria-hidden="true" tabindex="-1"></a>  <span class="kw">let</span> colors<span class="op">=</span>[<span class="st">"#EF740E"</span><span class="op">,</span><span class="st">"#EF740E"</span><span class="op">,</span><span class="st">"#B7040D"</span><span class="op">,</span><span class="st">"#B7040D"</span><span class="op">,</span><span class="st">"#48ABAB"</span><span class="op">,</span><span class="st">"#48ABAB"</span>]</span>
<span id="cb53-986"><a href="#cb53-986" aria-hidden="true" tabindex="-1"></a>  <span class="kw">let</span> structures<span class="op">=</span>[]</span>
<span id="cb53-987"><a href="#cb53-987" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-988"><a href="#cb53-988" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (<span class="kw">let</span> [index<span class="op">,</span>file] <span class="kw">of</span> files<span class="op">.</span><span class="fu">entries</span>()){</span>
<span id="cb53-989"><a href="#cb53-989" aria-hidden="true" tabindex="-1"></a>    <span class="kw">let</span> structure <span class="op">=</span> <span class="cf">await</span> stage5<span class="op">.</span><span class="fu">loadFile</span>(file<span class="op">,</span> {<span class="dt">ext</span><span class="op">:</span> <span class="st">"pdb"</span><span class="op">,</span> <span class="dt">asTrajectory</span><span class="op">:</span> <span class="kw">false</span>})<span class="op">;</span></span>
<span id="cb53-990"><a href="#cb53-990" aria-hidden="true" tabindex="-1"></a>    structure<span class="op">.</span><span class="fu">addRepresentation</span>(<span class="st">"ribbon"</span><span class="op">,</span>{ <span class="st">"radius"</span><span class="op">:</span><span class="fl">0.1</span><span class="op">,</span><span class="st">"sele"</span><span class="op">:</span> <span class="st">"protein"</span><span class="op">,</span> <span class="st">"color"</span><span class="op">:</span> colors[index]})<span class="op">;</span></span>
<span id="cb53-991"><a href="#cb53-991" aria-hidden="true" tabindex="-1"></a>    structures<span class="op">.</span><span class="fu">push</span>(structure)</span>
<span id="cb53-992"><a href="#cb53-992" aria-hidden="true" tabindex="-1"></a>  }<span class="op">;</span></span>
<span id="cb53-993"><a href="#cb53-993" aria-hidden="true" tabindex="-1"></a>  structures[<span class="dv">0</span>]<span class="op">.</span><span class="fu">autoView</span>(<span class="st">"(106-107 or 112 or 131-139) and .CA"</span>)<span class="op">;</span></span>
<span id="cb53-994"><a href="#cb53-994" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-10-1" data-nodetype="declaration">

</div>
</div>
</div>
<div class="cell-output cell-output-display hidden">
<div>
<div id="ojs-cell-10-2" data-nodetype="expression">

</div>
</div>
</div>
</div>
<p>In the 3D View we can nicely see that we clustered the different conformations that occur for the majority of HSP90 alpha human structures in the PDB.</p>
<div class="cell" data-execution_count="15">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.cluster.hierarchy <span class="im">import</span> cut_tree</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>h_cluster_id <span class="op">=</span> cut_tree(clusters, height<span class="op">=</span><span class="fl">1.5</span>).reshape(<span class="op">-</span><span class="dv">1</span>, )</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>sns.displot(h_cluster_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div id="fig-cluster-histogram" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="2023-04-03-binding-site-comparison-benchmark-I_files/figure-html/fig-cluster-histogram-output-1.png" width="469" height="469" class="figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;13: Size distribution of identified clusters (cut at 1A distance).</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>The histogram above here shows nicely these 3 larger groups and the rest. Obviously we can play around with where do we cut the dendrogram to get more or less clusters. However, here Iâ€™ll rather stick to the current cut at 1A. We can revise that at a later stage if needed.</p>
</section>
<section id="wrap-up" class="level3">
<h3 class="anchored" data-anchor-id="wrap-up">Wrap-up</h3>
<p>Well that got more complicated than it should have been, didnâ€™t it? And unfortunately that was the easy part. Now we can extend the set gradually <span class="emoji" data-emoji="tada">ðŸŽ‰</span>.</p>
<p>But letâ€™s resume a bit what can be done with these results &amp; scripts I provide here. We now have a set of structures that: 1. have been automagically selected from the PDB 2. this selection can be updated at any time (unless they change their API, which they will at some point) 3. we have performed local structure comparison to know which known ATP binding sites are similar to each other in terms of conformation</p>
<p>This set of structures, and especially the clusters, can be used to serve as known positives for binding site benchmarking. It can also be used to evaluate rank ordering for binding site comparison methods.</p>
<div class="cell" data-execution_count="16">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> Markdown</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> tabulate <span class="im">import</span> tabulate</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>identifiers<span class="op">=</span>[<span class="st">"</span><span class="sc">{}</span><span class="st">:</span><span class="sc">{}</span><span class="st">"</span>.<span class="bu">format</span>(a_, b_) <span class="cf">for</span> a_, b_ <span class="kw">in</span> <span class="bu">zip</span>(filteredLabels, filteredChainCodes)]</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>table <span class="op">=</span> <span class="bu">list</span>(<span class="bu">zip</span>(identifiers, h_cluster_id))</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>Markdown(tabulate(</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  table, </span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>  headers<span class="op">=</span>[<span class="st">"Structure"</span>,<span class="st">"Cluster ID"</span>], colalign<span class="op">=</span>(<span class="st">"center"</span>,<span class="st">"center"</span>)</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="16">
<table class="table table-sm table-striped small">
<caption>Final List of structures and their corresponding cluster ID of the selected HSP90 alpha human ATP binding site containing structures </caption>
<thead>
<tr class="header">
<th style="text-align: center;">Structure</th>
<th style="text-align: center;">Cluster ID</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1BYQ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">1OSF:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1UY6:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">1UY7:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1UY8:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">1UY9:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1UYC:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">1UYD:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1UYE:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">1UYF:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1UYG:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">1UYH:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1UYI:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">1UYK:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1UYL:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">1YC1:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1YC3:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">1YC4:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1YER:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">1YES:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1YET:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2BSM:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2BT0:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2BYH:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2BYI:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2BZ5:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2CCS:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2CCT:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2CCU:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2FWY:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2FWZ:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">2H55:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2JJC:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2K5B:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2QF6:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2QFO:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2QG0:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2QG2:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2UWD:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2VCI:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2VCJ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2WI1:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2WI2:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2WI3:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2WI4:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">2WI5:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2WI6:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2WI7:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2XAB:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2XDK:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2XDL:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2XDS:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2XDU:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">2XDX:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2XHR:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2XHT:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2XHX:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2XJG:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2XJJ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2XJX:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2XK2:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YE2:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YE3:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YE4:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YE5:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YE6:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YE7:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YE8:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YE9:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YEA:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YEB:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YEC:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YED:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YEE:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YEF:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YEG:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YEH:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YEI:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YEJ:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YI0:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YI6:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YI7:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YJW:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YJX:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YK2:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YK9:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YKB:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YKC:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YKE:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">2YKI:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">2YKJ:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3B24:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3B25:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3B26:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3B27:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3B28:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3BM9:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3BMY:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3D0B:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3EKO:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3EKR:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3FT5:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3FT8:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3HEK:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3HHU:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3HYY:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3HZ1:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3HZ5:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3INW:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3INX:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3K97:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3K98:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3K99:A</td>
<td style="text-align: center;">2</td>
</tr>
<tr class="even">
<td style="text-align: center;">3MNR:P</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3O0I:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3OW6:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3OWB:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3OWD:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3QDD:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3QTF:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3R4M:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3R4N:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3R4O:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3R4P:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3R91:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3R92:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3RKZ:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">3RLP:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3RLQ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3RLR:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3T0H:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3T0Z:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3T10:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3T1K:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3T2S:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3TUH:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3VHA:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3VHC:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3VHD:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">3WHA:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3WQ9:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4AWO:A</td>
<td style="text-align: center;">3</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4AWP:A</td>
<td style="text-align: center;">3</td>
</tr>
<tr class="even">
<td style="text-align: center;">4AWQ:A</td>
<td style="text-align: center;">3</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4B7P:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4BQG:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4BQJ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4CWF:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4CWN:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4CWO:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4CWP:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4CWQ:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4CWR:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4CWS:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4CWT:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4EEH:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4EFT:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4EFU:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4EGH:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4EGI:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4EGK:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4FCP:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4FCQ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4HY6:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4JQL:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4L8Z:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4L90:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4L91:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4L93:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4L94:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4LWE:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4LWF:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4LWG:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4LWH:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4LWI:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4NH7:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4NH8:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4O04:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4O05:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4O07:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4O09:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4O0B:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4R3M:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4U93:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4W7T:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4XIP:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4XIQ:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4XIR:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4XIT:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">4YKQ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4YKR:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4YKT:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4YKU:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4YKW:A</td>
<td style="text-align: center;">4</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4YKX:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">4YKY:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">4YKZ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">5CF0:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5FNC:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5FND:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5FNF:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5GGZ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5J20:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5J27:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5J2V:A</td>
<td style="text-align: center;">5</td>
</tr>
<tr class="even">
<td style="text-align: center;">5J2X:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5J64:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">5J6L:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5J6M:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">5J6N:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5J82:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5J86:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5J8M:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">5J8U:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5J9X:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5LNY:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5LNZ:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5LO0:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5LO1:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5LO5:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5LO6:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">5LQ9:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5LR1:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5LR7:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5LRL:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5LRZ:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5LS1:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5M4E:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5M4H:A</td>
<td style="text-align: center;">2</td>
</tr>
<tr class="even">
<td style="text-align: center;">5NYH:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5NYI:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">5OCI:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5OD7:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5ODX:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5T21:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5VYY:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5XQD:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">5XQE:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5XR5:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5XR9:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5XRB:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5XRD:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">5XRE:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">5ZR3:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6B99:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6B9A:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6CEO:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6CYG:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6CYH:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6EI5:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6EL5:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">6ELN:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6ELO:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6ELP:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6EY8:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">6EY9:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6EYB:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">6F1N:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6FCJ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6GPR:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6GPT:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6GPW:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6GPY:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6GQ6:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6GQR:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6HHR:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6LR9:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">6LSZ:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6LT8:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6LTI:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6LTK:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">6N8X:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6TN4:AAA</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">6TN5:AAA</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">6U99:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">6U9B:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7DMC:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">7KRJ:A</td>
<td style="text-align: center;">6</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7KW7:A</td>
<td style="text-align: center;">2</td>
</tr>
<tr class="even">
<td style="text-align: center;">7L7I:A</td>
<td style="text-align: center;">6</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7L7J:A</td>
<td style="text-align: center;">6</td>
</tr>
<tr class="even">
<td style="text-align: center;">7LSZ:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7LT0:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">7S8Y:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7S8Z:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">7S90:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7S95:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">7S98:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7S99:A</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">7S9F:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7S9G:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">7S9H:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="odd">
<td style="text-align: center;">7S9I:A</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">7UR3:A</td>
<td style="text-align: center;">7</td>
</tr>
<tr class="odd">
<td style="text-align: center;">8AGI:A</td>
<td style="text-align: center;">8</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="comparison-to-existing-literature" class="level3">
<h3 class="anchored" data-anchor-id="comparison-to-existing-literature">Comparison to existing literature</h3>
<p>This section is to verify how this very first section on very similar binding sites differs from already existing papers. The idea is to provide constructive criticism and not to bash anyoneâ€™s work - and probably Iâ€™ll have to revise a few things on my side while reading yet again into some details of some of the published work.</p>
<section id="prospeccts-ehrt2018" class="level4">
<h4 class="anchored" data-anchor-id="prospeccts-ehrt2018">Prospeccts <span class="citation" data-cites="Ehrt2018">Ehrt, Brinkjost, and Koch (<a href="#ref-Ehrt2018" role="doc-biblioref">2018</a>)</span></h4>
<blockquote class="blockquote">
<p>The first data set (structures with identical sequences) was designed to evaluate the toolsâ€™ sensitivity to the binding site definition. This definition often depends on the size and location of bound ligands. Different ligands can address various regions of the binding site of interest (sub-pockets). Because of these different site definitions, similar sub-pockets are more difficult to match. Although binding sites can interact with a broad variety of ligands, they share common properties and distinct similarities. Thus, the scoring scheme has to be optimized for this scenario to enrich similar binding site pairs with different cavity definitions in a ranked list. Data set 1 contains structures with identical sequences, which bind to chemically different ligands located at identical sites, leading to diverse binding site definitions used for the comparisons.</p>
</blockquote>
<ul>
<li>First first major difference that pops up here is that the binding site appears to be defined using the ligand bound to HSP90
<ul>
<li>This is something I already discussed in my previous post on the drawbacks of binding site comparison benchmarks</li>
<li>No apo structures can be considered here</li>
<li>On top of structural variation one adds the variation of the definition on where the binding site exactly is
<ul>
<li><span class="citation" data-cites="Ehrt2018">Ehrt, Brinkjost, and Koch (<a href="#ref-Ehrt2018" role="doc-biblioref">2018</a>)</span> et al are totally right that this is something to check. Iâ€™m just relunctant to use this to define an ideal benchmark set.<br>
</li>
</ul></li>
</ul></li>
<li>Second they cluster by identical sequence, as opposed to what I did here (same protein strictly speaking)
<ul>
<li>that very likely something I might revise on my side when starting with the 2nd set of structures with similar sequences on non identical proteins - but for letâ€™s keep it clean &amp; simple.</li>
</ul></li>
</ul>
<blockquote class="blockquote">
<p>The sequences of all single chain entries of the sc-PDB[65] were sequence culled using the PISCES[104] server (with a sequence identity threshold of 25%). The single chain sc-PDB entry sequences of the culled set were compared to those of all single chain proteins as stored in the PDB using USEARCH[105]. Structures whose sequences were identical to those of at least nine others were retained. This resulted in a data set of 13 groups of diverse structures. The PDB structures within each group were aligned and the ligand-occupied binding sites were compared to ensure that all ligands were located at the same site. This visual inspection led to the exclusion of one group represented by the sc-PDB entry with the PDB ID 4l8u (chain A) as the ligand-defined binding sites differed for the structures with 100% sequence identity. A complete list of data set 1 can be found in S9 Table. An all-against-all comparison was performed to assess the performance of each tool. Site pairs of proteins with identical sequences were regarded as positives while other matches were defined as negatives.</p>
</blockquote>
<!-- 
## SCOP classification

If one considers that the SCOP family of this domain will likely contain other HSP90 Nter ATP binding sites then this is a useful resource to look for already assigned similarities by protein fold classification. This has been used already in several other papers in the litterature, but let's exemplify here for the sake of completeness. 

The N-ter ATP binding domain of HSP90 and the hierarchy of [SCOP classifications can be found here](https://scop.mrc-lmb.cam.ac.uk/term/8028628).

![scop_hsp90.png](scop_hsp90.png)

 -->



</section>
</section>
</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Citri2006" class="csl-entry" role="listitem">
Citri, Ami, Daniel Harari, Galit Shohat, Parameswaran Ramakrishnan, Judith Gan, Sara Lavi, Miriam Eisenstein, et al. 2006. <span>â€œHsp90 Recognizes a Common Surface on Client Kinases.â€</span> <em>Journal of Biological Chemistry</em> 281 (20): 14361â€“69. <a href="https://doi.org/10.1074/jbc.m512613200">https://doi.org/10.1074/jbc.m512613200</a>.
</div>
<div id="ref-Damm2006" class="csl-entry" role="listitem">
Damm, Kelly L., and Heather A. Carlson. 2006. <span>â€œGaussian-Weighted <span>RMSD</span> Superposition of Proteins: A Structural Comparison for Flexible Proteins and Predicted Protein Structures.â€</span> <em>Biophysical Journal</em> 90 (12): 4558â€“73. <a href="https://doi.org/10.1529/biophysj.105.066654">https://doi.org/10.1529/biophysj.105.066654</a>.
</div>
<div id="ref-Dana2018" class="csl-entry" role="listitem">
Dana, Jose M, Aleksandras Gutmanas, Nidhi Tyagi, Guoying Qi, Claire Oâ€™Donovan, Maria Martin, and Sameer Velankar. 2018. <span>â€œ<span>SIFTS</span>: Updated Structure Integration with Function, Taxonomy and Sequences Resource Allows 40-Fold Increase in Coverage of Structure-Based Annotations for Proteins.â€</span> <em>Nucleic Acids Research</em> 47 (D1): D482â€“89. <a href="https://doi.org/10.1093/nar/gky1114">https://doi.org/10.1093/nar/gky1114</a>.
</div>
<div id="ref-Ehrt2018" class="csl-entry" role="listitem">
Ehrt, Christiane, Tobias Brinkjost, and Oliver Koch. 2018. <span>â€œA Benchmark Driven Guide to Binding Site Comparison: An Exhaustive Evaluation Using Tailor-Made Data Sets (<span>ProSPECCTs</span>).â€</span> Edited by Bjoern Peters. <em><span>PLOS</span> Computational Biology</em> 14 (11): e1006483. <a href="https://doi.org/10.1371/journal.pcbi.1006483">https://doi.org/10.1371/journal.pcbi.1006483</a>.
</div>
<div id="ref-Kabsch1976" class="csl-entry" role="listitem">
Kabsch, W. 1976. <span>â€œA Solution for the Best Rotation to Relate Two Sets of Vectors.â€</span> <em>Acta Crystallographica Section A</em> 32 (5): 922â€“23. <a href="https://doi.org/10.1107/s0567739476001873">https://doi.org/10.1107/s0567739476001873</a>.
</div>
<div id="ref-Kufareva2011" class="csl-entry" role="listitem">
Kufareva, Irina, and Ruben Abagyan. 2011. <span>â€œMethods of Protein Structure Comparison.â€</span> In <em>Methods in Molecular Biology</em>, 231â€“57. Humana Press. <a href="https://doi.org/10.1007/978-1-61779-588-6_10">https://doi.org/10.1007/978-1-61779-588-6_10</a>.
</div>
<div id="ref-Soti2003" class="csl-entry" role="listitem">
Soti, Csaba, Akos Vermes, Timothy A. J. Haystead, and Peter Csermely. 2003. <span>â€œComparative Analysis of the <span>ATP</span>-Binding Sites of Hsp90 by Nucleotide Affinity Cleavage: A Distinct Nucleotide Specificity of the c-Terminal <span>ATP</span>-Binding Site.â€</span> <em>European Journal of Biochemistry</em> 270 (11): 2421â€“28. <a href="https://doi.org/10.1046/j.1432-1033.2003.03610.x">https://doi.org/10.1046/j.1432-1033.2003.03610.x</a>.
</div>
<div id="ref-Velankar2012" class="csl-entry" role="listitem">
Velankar, Sameer, JosÃ© M. Dana, Julius Jacobsen, Glen van Ginkel, Paul J. Gane, Jie Luo, Thomas J. Oldfield, Claire Oâ€™Donovan, Maria-Jesus Martin, and Gerard J. Kleywegt. 2012. <span>â€œ<span>SIFTS</span>: Structure Integration with Function, Taxonomy and Sequences Resource.â€</span> <em>Nucleic Acids Research</em> 41 (D1): D483â€“89. <a href="https://doi.org/10.1093/nar/gks1258">https://doi.org/10.1093/nar/gks1258</a>.
</div>
</div></section></div></main> <!-- /main -->
<script type="ojs-module-contents">
{"contents":[{"methodName":"interpret","cellName":"ojs-cell-1","inline":false,"source":"// Create drawing area\ndivNGL = html`<div style=\"width:800px;height:600px;position:relative\"></div>`;\n"},{"methodName":"interpret","cellName":"ojs-cell-2","inline":false,"source":"NGL = require(\"ngl@next\");\n// Create trajectory object\nstage = new NGL.Stage(divNGL, { backgroundColor: \"black\",cameraType: \"orthographic\"});\npdbString = await FileAttachment(\"3t0z.pdb\").blob();\nstructure = await stage.loadFile(pdbString, {ext: \"pdb\", asTrajectory: false})\nstructure.addRepresentation(\"cartoon\",{ \"sele\": \"protein\", \"color\": \"residueindex\"});\nstructure.addRepresentation(\"ball+stick\",{ \"sele\": \"hetero and _C\", \"color\": \"yellow\"});\nstructure.addRepresentation(\"ball+stick\",{ \"sele\": \"hetero\", \"color\": \"element\"});\nstructure.addRepresentation(\"licorice\",{ \"sele\": \"47-55 or 91 or 93 or 96-97 or 98 or 102 or 106 or 107 or 112 or 132-139 or 150 or 162 or or 186\", \"color\": \"element\"});\nstructure.addRepresentation(\"contact\",{ \"sele\": \"47-55 or 91 or 93 or 96-97 or 98 or 102 or 106 or 107 or 112 or 132-139 or 150 or 162 or or 186 or ligand or water\", \"color\": \"element\"});\nstructure.autoView();\n"},{"methodName":"interpret","cellName":"ojs-cell-3","inline":false,"source":"// Create drawing area\ndivNGL2 = html`<div style=\"width:800px;height:600px;position:relative\"></div>`;\n"},{"methodName":"interpret","cellName":"ojs-cell-4","inline":false,"source":"stage2 = new NGL.Stage(divNGL2, { backgroundColor: \"black\",cameraType: \"orthographic\"});\npdbString1 = await FileAttachment(\"3t0z.pdb\").blob();\nstructure1 = await stage2.loadFile(pdbString1, {ext: \"pdb\", asTrajectory: false})\nstructure1.addRepresentation(\"cartoon\",{ \"sele\": \"protein\", \"color\": \"#325E73\"});\nstructure1.addRepresentation(\"ball+stick\",{ \"sele\": \"ATP\", \"color\": \"element\"});\npdbString2 = await FileAttachment(\"pdbs/4cwr_aligned.pdb\").blob();\nstructure2 = await stage2.loadFile(pdbString2, {ext: \"pdb\", asTrajectory: false})\nstructure2.addRepresentation(\"cartoon\",{ \"sele\": \"protein\", \"color\": \"#B69849\"});\npdbString3 = await FileAttachment(\"pdbs/7l7j_aligned.pdb\").blob();\nstructure3 = await stage2.loadFile(pdbString3, {ext: \"pdb\", asTrajectory: false})\nstructure3.addRepresentation(\"cartoon\",{ \"sele\": \"protein\", \"color\": \"#B66049\"});\n\nstructure1.autoView(\"ATP\");\n"},{"methodName":"interpret","cellName":"ojs-cell-5","inline":false,"source":"// Create drawing area\ndivNGL3 = html`<div style=\"width:800px;height:600px;position:relative\"></div>`;\n"},{"methodName":"interpret","cellName":"ojs-cell-6","inline":false,"source":"stage3 = new NGL.Stage(divNGL3, { backgroundColor: \"black\",cameraType: \"orthographic\"});\npdbString31 = await FileAttachment(\"3t0z.pdb\").blob();\nstructure31 = await stage3.loadFile(pdbString31, {ext: \"pdb\", asTrajectory: false})\nstructure31.addRepresentation(\"trace\",{ \"sele\": \"protein\", \"color\": \"#325E73\"});\nstructure31.addRepresentation(\"ball+stick\",{ \"sele\": \"(106-107 or 112 or 131-139) and .CA\", \"color\": \"#325E73\"});\nstructure31.addRepresentation(\"label\", {\"sele\": \"(106-107 or 112 or 131-139) and .CA\", \"labelType\":\"resno\", \"color\": \"#7F9DAC\"});\n\npdbString32 = await FileAttachment(\"pdbs/4cwr_aligned.pdb\").blob();\nstructure32 = await stage3.loadFile(pdbString32, {ext: \"pdb\", asTrajectory: false});\nstructure32.addRepresentation(\"trace\",{ \"sele\": \"protein\", \"color\": \"#B69849\"});\nstructure32.addRepresentation(\"ball+stick\",{ \"sele\": \"(106-107 or 112 or 131-139) and .CA and :A\", \"color\": \"#B69849\"});\nstructure32.addRepresentation(\"label\", {\"sele\": \"(106-107 or 112 or 131-139) and .CA and :A\", \"labelType\":\"resno\", \"color\": \"#FFEBB8\"});\nstructure31.autoView(\"(106-107 or 112 or 131-139) and .CA and :A\");\n"},{"methodName":"interpret","cellName":"ojs-cell-7","inline":false,"source":"// Create drawing area\ndivNGL4 = html`<div style=\"width:800px;height:600px;position:relative\"></div>`;\n"},{"methodName":"interpret","cellName":"ojs-cell-8","inline":false,"source":"stage4 = new NGL.Stage(divNGL4, { backgroundColor: \"black\",cameraType: \"orthographic\"});\n\npdbString42 = await FileAttachment(\"pdbs/4cwr_aligned.pdb\").blob();\nstructure42 = await stage4.loadFile(pdbString42, {ext: \"pdb\", asTrajectory: false});\nstructure42.addRepresentation(\"trace\",{ \"sele\": \"protein\", \"color\": \"#B69849\"});\nstructure42.addRepresentation(\"ball+stick\",{ \"sele\": \"(106 or 112 or 131-133) and .CA\", \"color\": \"#B69849\"});\nstructure42.addRepresentation(\"label\", {\"sele\": \"(106 or 112 or 131-133) and .CA\", \"labelType\":\"resno\", \"color\": \"#FFEBB8\"});\npdbString41 = await FileAttachment(\"pdbs/7l7j_aligned.pdb\").blob();\nstructure41 = await stage4.loadFile(pdbString41, {ext: \"pdb\", asTrajectory: false});\nstructure41.addRepresentation(\"trace\",{ \"sele\": \"protein\", \"color\": \"#B66049\"});\nstructure41.addRepresentation(\"ball+stick\",{ \"sele\": \"(106 or 112 or 131-133) and .CA and :A\", \"color\": \"#B66049\"});\nstructure41.addRepresentation(\"label\", {\"sele\": \"(106 or 112 or 131-133) and .CA and :A\", \"labelType\":\"resno\", \"color\": \"#FFC7B8\"});\nstructure41.autoView(\"(106 or 112 or 131-133) and .CA and :A\");\n\n"},{"methodName":"interpret","cellName":"ojs-cell-9","inline":false,"source":"// Create drawing area\ndivNGL5 = html`<div style=\"width:800px;height:600px;position:relative\"></div>`;\n"},{"methodName":"interpret","cellName":"ojs-cell-10","inline":false,"source":"stage5 = new NGL.Stage(divNGL5, { backgroundColor: \"black\",cameraType: \"orthographic\"});\n{\n  let files=[await FileAttachment(\"pdbs/4yku_aligned.pdb\").blob(),\n    await FileAttachment(\"pdbs/2yi6_aligned.pdb\").blob(),\n    await FileAttachment(\"pdbs/1yc4_aligned.pdb\").blob(),\n    await FileAttachment(\"pdbs/2xjg_aligned.pdb\").blob(),\n    await FileAttachment(\"pdbs/4o05_aligned.pdb\").blob(),\n    await FileAttachment(\"pdbs/2xdu_aligned.pdb\").blob()]\n  let colors=[\"#EF740E\",\"#EF740E\",\"#B7040D\",\"#B7040D\",\"#48ABAB\",\"#48ABAB\"]\n  let structures=[]\n\n  for (let [index,file] of files.entries()){\n    let structure = await stage5.loadFile(file, {ext: \"pdb\", asTrajectory: false});\n    structure.addRepresentation(\"ribbon\",{ \"radius\":0.1,\"sele\": \"protein\", \"color\": colors[index]});\n    structures.push(structure)\n  };\n  structures[0].autoView(\"(106-107 or 112 or 131-139) and .CA\");\n}\n"}]}
</script>
<script type="module">
if (window.location.protocol === "file:") { alert("The OJS runtime does not work with file:// URLs. Please use a web server to view this document."); }
window._ojs.paths.runtimeToDoc = "../../wip/2023-04-03-binding-site-comparison-benchmark-I";
window._ojs.paths.runtimeToRoot = "../..";
window._ojs.paths.docToRoot = "../..";
window._ojs.selfContained = false;
window._ojs.runtime.interpretFromScriptTags();
</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'alternate';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>